---
title: "Extras"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Extras}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This vignette explores more thoroughly some aspects of our work that might be overlooked in the other vignettes or in our paper. In particular, we will

- Explain how we generate TAD boundaries for analyzing 10X Single Cell ATAC-seq data 
- Demonstrate how our method works on data involving only independent sets of pairwise distances (see Section 4.1 of our paper)
- Provide a population genetics example of how to run our method from the terminal

If you've come here for the details, you're in the right place ðŸ™‚

# Generating Topologically Associating Domains for scATAC-seq Data Analysis

*This Section supplements our [Analysis of Single Cell ATAC-seq Data](https://alanaw1.github.io/flintyR/articles/single-cell-atac-seq.html) vignette.*

## What are Topologically Associating Domains (TADs)?  

*Adapted from two review papers, [Dixon et al. (2016)](https://doi.org/10.1016/j.molcel.2016.05.018) appearing in Molecular Cell and [Szabo et al. (2019)](https://doi.org/10.1126/sciadv.aaw1668) appearing in Science Advances.*

To understand TADs, we must first understand how a eukaryotic genome organises itself in space. Scientists already proposed, more than one hundred years ago, that chromosomes in "resting" interphase nuclei (i.e., not undergoing division) exist as chromosome territories. The figure below shows the writing of the late 19th century German zoologist Theodor Boveri, in which he described the looping of chromosomes amongst other ideas he had about "chromosome individuality."            

<p align="center">
<img src="boveri_p202.png" width="400"/>
</p>
<p align="center">
*Obtained from [Boveri (1909)](https://www.biodiversitylibrary.org/page/52621809), available freely through the Biodiversity Heritage Library.*
</p>

Similar to how chromosomes in the genome organise themselves territorially, each chromosome also has its own organisation, also referred to as chromosome folding. Within each chromosome are histone proteins, around which DNA is wound. Such compact DNA-histone complexes are called chromatin. The degree of packing of these complexes (i.e., how tightly compact or loose) is a dynamic process that depends on the stage of the cell cycle the cell is at.

Recent technologies have allowed the quantification of the folding patterns described above. Notably, chromosome conformation capture methods like Hi-C ([Lieberman-Aiden et al., 2009](https://doi.org/10.1126/science.1181369)) have uncovered folding patterns across different scales, measured through interactions (chromosomal contact frequencies). Here are two notable patterns:

- At large scales, chromosomes segregate into regions of preferential long-range interactions, forming two mutually excluded types of chromatin. These types correspond to gene-rich and active chromatin and to repressive chromatin.  
- At a scale of tens to hundreds of kilobases, chromosomes fold into domains with preferential intradomain interactions compared to interdomain interactions. These contact domains are called TADs. 

The figure below provides a visualisation of these folding patterns. 

<p align="center">
<img src="szabo_fig1.jpg" width="400"/>
</p>
<p align="center">
*Hierarchical folding of the eukaryotic genome. (A) Many levels of chromatin folding, from the finest (DNA-histone association) to the coarsest (segregation of chromatin into active and repressed compartments). Each chromosome occupies its own territory within the nucleus. (B) Schematic representation of Hi-C maps at different genomic scales, reflecting different layers of high-order chromosome folding. Genomic coordinates are indicated along both axes. Figure reproduced from [Szabo et al. (2019)](https://advances.sciencemag.org/content/5/4/eaaw1668) under the Creative Commons Attribution NonCommercial (CC-BY-NC) license.*
</p>

## How do we find TADs?

Hi-C experiments measure contact frequencies across the genome, thus allowing scientists to identify TADs, which are the chromatin regions more frequently interacting within themselves than among each other. Before we discuss quantitative methods, it will be useful to see what our data looks like. Below, we load Hi-C contact maps obtained from human fibroblast cells (IMR90), as reported in [Dixon et al. (2012)](https://www.nature.com/articles/nature11082). The reason our data is named `hic_imr90_40` is that the Hi-C contact reads are binned at $40,000$ bp resolution; this means that read counts are combined across genomic regions of length $40,000$ bp. 

```{r hi_c_1, echo = TRUE, eval = TRUE, warning = FALSE, message = FALSE}
## Load libraries 
library(tidyverse)
require(HiTC) # install using Bioconductor if not installed
require(HiCDataHumanIMR90) # install using Bioconductor if not installed

## Load data
data(Dixon2012_IMR90) 

## Preview data
# What type of object is it, and how large?
class(hic_imr90_40) 
object.size(hic_imr90_40)

# What does a Hi-C map look like?
hic_imr90_40$chr1chr1@intdata[11:20,11:20]

# How are Hi-C features labeled?
head(hic_imr90_40$chr1chr1@xgi)
```

For the rest of this vignette we restrict to Chromosomes 1 and 2, consistent with what we report in our paper. As part of pre-processing, we perform the following steps, as described in Sections 4 and 5.3 of the **HiTC** [vignette](https://bioconductor.org/packages/release/bioc/vignettes/HiTC/inst/doc/HiC_analysis.pdf).

- Binning contact frequency measurements into windows of size $500,000$ bp (i.e., *reducing* the resolution from the original $40,000$ bp)
- Normalising binned contact frequencies with Iterative Correction and Eigenvector decomposition (ICE)

```{r hi_c_2, echo = TRUE, eval = TRUE, warning = FALSE, message = FALSE, out.width = '50%', cache = TRUE}
## Restrict Hi-C data to chr1 and chr2
sset <- reduce(hic_imr90_40, chr=c("chr1", "chr2"))

## Bin contact frequencies
imr90_500 <- HTClist(mclapply(sset, binningC, 
                              binsize=500000, 
                              bin.adjust=FALSE, 
                              method="sum",
                              step = 1))

## Perform ICE normalization
imr90_500_ICE <-  normICE(imr90_500, max_iter=1000) # converged at iteration 764

## Visualise transformed output of contact map
mapC(HTClist(imr90_500_ICE$chr1chr1), trim.range=.95,
     col.pos=c("white", "orange", "red", "black"))
```

We are now ready to identify TADs on our pre-processed dataset. As mentioned earlier, TADs are distinguished by intra-regional increase in interactions and inter-regional scarcity of interactions. We see this in the heat map above and can heuristically define TADs by inspection. However, to identify TADs in an automated fashion, we need systematic algorithms.

Here, we introduce two such systematic approaches. The first approach, called **domainCaller**, uses the Directionality Index (DI) to call TADs. Let the set of all $40,000$ bp resolution counts be $\mathscr{X}$. For each bin of contact frequencies $x\in\mathscr{X}$, let 
\begin{equation*}
\text{DI}(x) = \left(\frac{B(x) - A(x)}{|B(x) - A(x)|}\right) \left(\frac{(A(x) - E(x))^2}{E(x)} + \frac{(B(x) - E(x))^2}{E(x)}\right),
\end{equation*}
where

- $A(x)$ is the number of reads that map from $x$ to the upstream $2$Mb
- $B(x)$ is the number of reads that map from $x$ to the downstream $2$Mb
- $E(x) = \frac{1}{2}(A(x) + B(x))$ is the expected number of reads under null of *no directional bias*

What **domainCaller** does is the following.

1. Perform DI computation across all bins $x\in\mathscr{X}$, obtaining a set of DIs $\{\text{DI}(x):x\in \mathscr{X}\}$.
2. Fit a hidden Markov model (HMM) to $\{\text{DI}(x):x\in \mathscr{X}\}$.
3. Use the inferred HMM states to define regions and classify them into TADs.

In Step 2 and Step 3, the HMM used is a three-state model (upstream bias, downstream bias, and no bias) with transition probabilities between successive bins. Details of this HMM can be found in Supplementary Figure 28 of [Dixon et al. (2012)](https://www.nature.com/articles/nature11082). 

One subtlety about TAD calling in Step 3: rather than inferring domains only, *both* domains and boundaries are inferred. A domain is initiated at the beginning of a single downstream biased state, and ends with the last state that is upstream biased. Regions between domains are classified as topological boundaries.  

Fortunately, **Hi-TC** has made available the output of running **domainCaller** on our dataset. This is useful for us to gain intuition about what TADs look like, in preparation for the second approach we will introduce. Below, we load the **domainCaller** output.

```{r hi_c_3, echo = TRUE, eval = TRUE, warning = FALSE, message = FALSE}
## Inspect domainCaller output
tads_imr90@ranges

## Inspect number of TADs called in each chromosome
tads_imr90@seqnames
```

The second approach to calling TADs is **TopDom**. Proposed by [Shin et al. (2016)](https://doi.org/10.1093/nar/gkv1505), **TopDom** relies on a bin signal statistic to identify TADs. For each bin $x\in\mathscr{X}$, let
\begin{equation*}
\text{binSignal}(x) = \frac{1}{w^2} \sum_{\ell=1}^w \sum_{m=1}^w \text{cont.freq}(U_x(\ell), U_x(m)),
\end{equation*}
where 

- $w$ is a window size parameter, to be fixed prior to identifying TAD boundaries 
- $U_x(\ell)$ and $D_x(m)$ are bins located $\ell$ indices upstream and $m$ indices downstream of bin $x$
- $\text{cont.freq}$ is the number of reads 

As the formula suggests, $\text{binSignal}$ computes the average contact frequency between upstream and downstream regions. See the Figure below for a visualisation.   

<p align="center">
<img src="shin_fig1A.png" width="400"/>
</p>
<p align="center">
*Denoting the bin $x$ by its index $i$, $\text{binSignal}(i)$ is the average contact frequency between an upstream and a downstream chromatin region ($U_i$ and $D_i$) in a window of size $2w$ surrounding the bin. Its value is relatively high if bin $i$ is located inside a TAD (red diamond), and reaches a local minimum at a TAD boundary. Figure adapted from [Shin et al. (2016)](https://doi.org/10.1093/nar/gkv1505) under the Creative Commons Attribution NonCommercial (CC-BY-NC) license.*
</p>

After computing bin signals for each bin, there is still a need to distinguish true local minima from false minima. To accomplish this, a sequential polygonal approximation technique is used to fit a smooth curve to the binsignal vs bin index graph. See the Figure below for a visualisation.

<p align="center">
<img src="shin_fig1B.png" width="400"/>
</p>
<p align="center">
*Turning points (blue circles) are identified in the original curve (black), and dominant local minima (red inverted triangles) are detected by fitting a piecewise linear curve (sequential polygonal approximation). Figure adapted from [Shin et al. (2016)](https://doi.org/10.1093/nar/gkv1505) under the Creative Commons Attribution NonCommercial (CC-BY-NC) license.*
</p>

Finally, to assign TAD boundaries, an additional Wilcoxon rank sum test is performed on the local minima selected. 

Below, we demonstrate performing these steps on Chromosome 1 of the IMR90 dataset. We use a recent packaged version of **TopDom**, provided by [Bengtsson et al. (2020)](https://github.com/HenrikBengtsson/TopDom).   

```{r hi_c_4, echo = TRUE, eval = TRUE, warning = FALSE, message = FALSE}
## Load package
# install.packages("TopDom")
library(TopDom)

## Convert to format readable by TopDom::TopDom
# 1. Make imr90_500_ICE into N x (N + 3) dataframe
# We use the multiple assignment function, %<-%, provided by zeallot
library(zeallot)
df <- data.frame(chr = character(),
                 start = numeric(),
                 end = numeric())
for (i in 1:length(colnames(imr90_500_ICE$chr1chr1@intdata))) {
  c(chr_, start_, end_) %<-% 
    strsplit(colnames(imr90_500_ICE$chr1chr1@intdata)[i], "\\:|\\-")[[1]]
  df <- rbind(df, 
              data.frame(chr = chr_,
                         start = as.numeric(start_),
                         end = as.numeric(end_)))
}
topdom_input_500k <- cbind(df, as.matrix(imr90_500_ICE$chr1chr1@intdata))
rownames(topdom_input_500k) <- c()
colnames(topdom_input_500k) <- NULL
topdom_input_500k[,2] <- topdom_input_500k[,2] - 1

# What's the dimension of the newly formatted object?
dim(topdom_input_500k) # it's N x (N+3)

# What does it look like? 
topdom_input_500k[1:6,1:10]

# 2. Save as tab-delimited file
rel_dir <- "ex_data/scATAC-seq/"
#write.table(topdom_input_500k, file = paste0(rel_dir, "500k_topdom_input.tsv"), 
#            row.names = FALSE, 
#            col.names = FALSE,
#            sep="\t")

# 3. Load the file and find TADs with TopDom
# TopDom performs the steps described above
topdom_res <- TopDom::TopDom(paste0(rel_dir, "500k_topdom_input.tsv"), window.size = 5L)

# 4. View results
# What are the p-values? 
head(topdom_res$binSignal)

# Which blocks are classified as TADs by TopDom?
head(topdom_res$bed[which(topdom_res$bed$name ==  "domain"), ])
```

To conclude this section, we look at the distribution of classification categories (not just TADs), as well as the distribution sizes (block lengths) of TADs. 

```{r hi_c_5, echo = TRUE, eval = TRUE, warning = FALSE, message = FALSE}
## Look at distribution of classification results
kableExtra::kable(table(topdom_res$bed$name),
                  col.names = c("Category",  "Count"))

## Look at distribution of size of TADs
chr1_tads <- topdom_res$domain[which(topdom_res$domain$tag =="domain"), ]
kableExtra::kable(table(chr1_tads$size / 5e5),
                  col.names = c("Size (x 5e5 bases)", "Count"))
```

Note we use these classified TADs to group our scATAC-seq features. 

# Test of Exchangeability Using Pairwise Distances

Our test of exchangeability works for settings where only pairwise distance data is available. Given a list of pairwise distances between $N$ individuals, $\{D_1,\ldots,D_B\}$, and assuming these distances are independent of one another, our test can determine whether the individuals making up the $N$-sample are exchangeable at a user-specified significance threshold (e.g., $\alpha=0.05$). (See Section of our paper.)

There are multiple practical scenarios well suited for such an approach.

- Loading the sample-by-feature matrix $\mathbf{X}$ into memory is computationally unfeasible
- Only distance data is available owing to privacy issues 
- Independent, large sets of features exist on different computers but need to be combined to form the sample-by-feature matrix $\mathbf{X}$       

Below, we show how our method can be called on a list of distances, provided the features from which the pairwise distances are derived are mutually independent. Briefly, we have $N=100$ individuals from whom we obtain $50$ pairwise distance data, $\{D_1,\ldots,D_{50}\}$. Each pairwise distance matrix $D_b$ is computed from a block of $10$ features, and we assume each block is independent of any other block.  

```{r dist_list_1, echo = TRUE, eval = TRUE, warning = FALSE, message = FALSE}
## Simulation of list of distances
# Set seed for reproducibility
set.seed(2021)

# Generate data
N <- 100
num_rows <- choose(N,2)
B <- 50
dist_list <- list()
for (b in 1:B) {
  # Generate fake data 
  if (b <= B/2) {
    fake_samples <- replicate(10, rbinom(N,1,0.5))
    fake_dist <- as.matrix(dist(fake_samples, method = "manhattan"))
  } else {
    fake_samples <- replicate(10, rnorm(N, sd = 10))
    fake_dist <- as.matrix(dist(fake_samples, method = "euclidean"))
  }
  
  # Add to list
  dist_list[[b]] <- fake_dist
}

# Clear garbage
gc()
```

We now run our test of exchangeability on the list of distances. 

```{r dist_list_2, echo = TRUE, eval = TRUE, warning = FALSE, message = FALSE}
## Apply test of exchangeability to simulated distance list
library(flintyR)
library(doParallel)
# Register parallel computation
registerDoParallel()

# Compute p-value
distDataPValue(dist_list) # should be >= 0.05
```

# Running Our Test from Terminal

For users who prefer to run jobs from the terminal, we describe how to embed our test within a script that can be executed immediately on the terminal. To facilitate our exposition, we use data from the 1000 Genomes (1KG) project ([The 1000 Genomes Project Consortium, 2015](https://www.nature.com/articles/nature15393)). The 1KG data consists of individual genotypes spanning 26 populations and 5 superpopulations. You may learn more about, and see some preliminary analyses of, the data [here](https://alanaw1.github.io/post/2021/03/03/visualizing-1000-genomes-data/). If you are a geneticist, you probably already know more interesting analyses of this data.     

The challenge here is that genomes are typically provided in VCF or BIM/BED/FAM or PGEN/PVAR/PSAM format. Moreover, they can be very large and thus unreadable.    

Suppose our task at hand is to run a test of exchangeability on each 1KG population. There are millions of polymorphic variants, and we assume variants lying in different chromosomes are independent of one another. Since there are 22 autosomal chromosomes, this means that our features are grouped into 22 large blocks.         

To accomplish our task, we perform the following steps. 

1. For each population, split the genome into the 22 autosomes. 
2. Per autosome:
    - load it into R and construct its distance matrix
    - remove the loaded autosome object from memory
3. Run test of exchangeability on list of distance matrices. (See previous Section on this.)  

<details>
  <summary><b>Preparing the population-level files</b></summary>
    To obtain population-level files (BIM/BED/FAM) that can be fed into Step 1 above, we perform the following steps: (1) Download 1KG Phase 3 data (PGEN and PVAR) from [here](https://www.cog-genomics.org/plink/2.0/resources), see **Merged dataset**; (2) Filter to keep only variants that are in HapMap3 and UK Biobank; (3) Remove related individuals; (4) Filter to keep only biallelic SNPs with minor allele count at least $1$ (*this differs from usual filtering of alleles that satisfy MAF at least $0.01$*); (5) Generate BED/BIM/FAM files for each population using population ID files. All of these steps are accomplished using PLINK2.    
</details> 

To be updated.
