<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction â€¢ flintyR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction">
<meta property="og:description" content="flintyR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">flintyR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/single-cell-atac-seq.html">Analysis of scATAC-seq Data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction</h1>
            
      
      
      <div class="hidden name"><code>intro.Rmd</code></div>

    </div>

    
    
<p><strong>flinty</strong> provides scientists a way to test if their datasets contain multiple subpopulations. This is an important question if the scientist wishes to infer quantities of interest from the dataset, for instance population-level parameters, without overlooking any latent heterogeneities.</p>
<p>Consider the following examples commonly encountered by practitioners.</p>
<ul>
<li>In population genetics, genetic samples are collected from individuals, say in a country like Mexico whose population history records admixture events and endogamous mating by cultural ancestry. It is not surprising if a dataset contains individuals that belong to distinguishable genetic communities. Unfortunately, absent information like family pedigrees, we might not have external means of grouping the individuals. This leads to less fine-scale analysis of the data and missed inferences about subpopulations.</li>
<li>In single cell genomics, unique molecular identifiers (UMIs) are collected for a large sample of cells, and clustering algorithms are run to generate homogeneous groups of cells. External validation is used to guide correctness of clustering, and the resulting groups are assumed homogeneous for downstream analysis. However, the clustering algorithm does not <em>provably</em> recover homogeneous clusters without mathematical justification and underlying modeling assumptions made about the data. This has prompted the use of <a href="https://doi.org/10.1080/01621459.2020.1833888">statistical hypothesis testing</a> to validate post-clustering homogeneity.<br>
</li>
<li>In survey questionnaires typical of the social sciences, the multivariate dataset <span class="math inline">\(\mathbf{X}\)</span> conceals latent factors that can be recovered statistically. Concretely, it is assumed that the <span class="math inline">\(P\)</span>-variate samples <span class="math inline">\(\mathbf{x}_1,\ldots,\mathbf{x}_N\)</span> making up <span class="math inline">\(\mathbf{X}\)</span> satisfy <span class="math display">\[\mathbf{x}_i = \mathbf{\Lambda}\boldsymbol{\xi}_i + \boldsymbol{\delta}_i\]</span> for each sample <span class="math inline">\(i\)</span>. Here <span class="math inline">\(\mathbf{\Lambda}\)</span> is the <span class="math inline">\(P\times M\)</span> matrix of loadings, <span class="math inline">\(\boldsymbol{\xi}_i\)</span> is the <span class="math inline">\(M\)</span>-dimensional sample-specific measurement at those <span class="math inline">\(M\)</span> factors, and <span class="math inline">\(\boldsymbol{\delta}_i\)</span> is a <span class="math inline">\(P\)</span>-dimensional residual vector. After learning the latent factors from the dataset (i.e., <span class="math inline">\(\mathbf{\Lambda}\)</span>, and by extension <span class="math inline">\(\boldsymbol{\xi}_i\)</span> and <span class="math inline">\(\boldsymbol{\delta}_i\)</span> for each <span class="math inline">\(i\)</span>), it is important to <a href="https://doi.org/10.1207/S15327906MBR3801_5">check that</a> the residual vectors <span class="math inline">\(\mathbf{\Delta}=[\boldsymbol{\delta}_1,\ldots,\boldsymbol{\delta}_N]^T\)</span> look independent and identically distributed, or at least exchangeable, so that no other latent factor was missed.</li>
</ul>
<p>For each example above, methods have been developed to address the issue described. These methods vary in their degree of success. We won't discuss the successes and limitations of such methods here. We merely want to suggest connections of our methods with some of these settings.</p>
<p>Instead, in this vignette we shall describe our general method of exchangeability and demonstrate its use on simulated data. Along the way, we will delight ourselves with a few morsels of data analysis, including <em>double dipping</em> and <em>runs tests</em>.</p>
<div id="setup" class="section level1">
<h1 class="hasAnchor">
<a href="#setup" class="anchor"></a>Setup</h1>
<p>This section to be updated.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">flintyR</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">doParallel</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/doParallel/man/registerDoParallel.html">registerDoParallel</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div id="basic-example" class="section level1">
<h1 class="hasAnchor">
<a href="#basic-example" class="anchor"></a>Basic Example</h1>
<p>We start simple. This example is motivated by recent work of Lucy Gao and Daniela Witten, which they presented in the International Seminar on Selective Inference and in the R/Medicine 2020 Virtual Conference respectively.</p>
<p>We simulate <span class="math inline">\(\mathbf{X}\)</span>, a dataset consisting of <span class="math inline">\(N=100\)</span> samples, where each sample consists of <span class="math inline">\(P=50\)</span> independent features, each feature a realization of a standard Gaussian random variable. Mathematically, the entries <span class="math inline">\(x_{np}\overset{\text{iid}}{\sim} N(0,1)\)</span>.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Generate some data</span>
<span class="co"># Set seed for reproducibility</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">420</span><span class="op">)</span>
<span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">5000</span><span class="op">)</span>,
            nrow <span class="op">=</span> <span class="fl">100</span>,
            ncol <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></code></pre></div>
<p>Under such settings <span class="math inline">\(\mathbf{X}\)</span> is homogeneous and the samples <span class="math inline">\(\mathbf{x}_1,\ldots,\mathbf{x}_N\)</span> are exchangeable, but if we were handed <span class="math inline">\(\mathbf{X}\)</span> without knowing its generative mechanism completely, we might want to perform clustering as part of exploratory data analysis.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Cluster data</span>
<span class="va">X_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span>
<span class="va">X_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span><span class="op">(</span><span class="va">X_df</span>, centers <span class="op">=</span> <span class="fl">2</span>, nstart <span class="op">=</span> <span class="fl">25</span><span class="op">)</span>

<span class="co"># Assign cluster labels</span>
<span class="va">X_df</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="va">X_cluster</span><span class="op">$</span><span class="va">cluster</span>

<span class="co"># Perform PCA to visualize</span>
<span class="va">X_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/princomp.html">princomp</a></span><span class="op">(</span><span class="va">X_df</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span><span class="op">)</span>
<span class="va">PC1and2</span> <span class="op">&lt;-</span> <span class="va">X_pca</span><span class="op">$</span><span class="va">scores</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">PC1and2</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="va">X_cluster</span><span class="op">$</span><span class="va">cluster</span> </code></pre></div>
<p>By performing <span class="math inline">\(k\)</span>-means clustering with <span class="math inline">\(k=2\)</span> as shown above, we obtain a cluster assignment. We can visualize this clustering by performing PCA on the samples and plotting the projection of each sample onto its first two principal components.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Visualize clustering</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">PC1and2</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Comp.1</span>, y <span class="op">=</span> <span class="va">Comp.2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span>,
             shape <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">xlab</span><span class="op">(</span><span class="st">"PC1"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ylab</span><span class="op">(</span><span class="st">"PC2"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>colour<span class="op">=</span><span class="fu">guide_legend</span><span class="op">(</span>title<span class="op">=</span><span class="st">"Cluster"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"A"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="intro_files/figure-html/2_3-1.png" width="700"></p>
<p>Next, we perform <a href="https://en.wikipedia.org/wiki/Mann%E2%80%93Whitney_U_test">Wilcoxon rank sum tests</a> between the two populations for each feature, applying Bonferroni correction to control the family-wise error rate (FWER) at <span class="math inline">\(\alpha=0.05\)</span>. Recall that the Wilcoxon rank sum (aka, Mann-Whitney) test is a non-parametric two-sample test for difference in sample distributions.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Differential analysis</span>
<span class="co"># Split into two clusters </span>
<span class="va">X_df_1</span> <span class="op">&lt;-</span> <span class="va">X_df</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> 
<span class="va">X_df_2</span> <span class="op">&lt;-</span> <span class="va">X_df</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span> 

<span class="co"># Compute p-values</span>
<span class="va">p_val_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span> <span class="op">{</span>
  <span class="co"># Perform differential analysis on feature j</span>
  <span class="va">da_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/wilcox.test.html">wilcox.test</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">X_df_1</span><span class="op">[</span>,<span class="va">j</span><span class="op">]</span>, 
                           y <span class="op">=</span> <span class="va">X_df_2</span><span class="op">[</span>,<span class="va">j</span><span class="op">]</span><span class="op">)</span>
  <span class="co"># Print result if significant</span>
  <span class="va">p_val_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">p_val_vector</span>, <span class="va">da_result</span><span class="op">$</span><span class="va">p.value</span><span class="op">)</span>
  <span class="kw">if</span> <span class="op">(</span><span class="va">da_result</span><span class="op">$</span><span class="va">p.value</span> <span class="op">&lt;</span> <span class="fl">0.001</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"The p-value for feature "</span>, <span class="va">j</span>, <span class="st">" is "</span>, <span class="va">da_result</span><span class="op">$</span><span class="va">p.value</span>, <span class="st">".\n"</span>, 
        sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"This is significant at FWER = 0.05 after Bonferroni correction.\n"</span><span class="op">)</span>
  <span class="op">}</span>
<span class="op">}</span>
<span class="co">#&gt; The p-value for feature 8 is 0.0005319207.</span>
<span class="co">#&gt; This is significant at FWER = 0.05 after Bonferroni correction.</span>
<span class="co">#&gt; The p-value for feature 14 is 0.0001147439.</span>
<span class="co">#&gt; This is significant at FWER = 0.05 after Bonferroni correction.</span>
<span class="co">#&gt; The p-value for feature 22 is 7.798508e-06.</span>
<span class="co">#&gt; This is significant at FWER = 0.05 after Bonferroni correction.</span>
<span class="co">#&gt; The p-value for feature 24 is 3.191586e-07.</span>
<span class="co">#&gt; This is significant at FWER = 0.05 after Bonferroni correction.</span>
<span class="co">#&gt; The p-value for feature 35 is 8.154343e-05.</span>
<span class="co">#&gt; This is significant at FWER = 0.05 after Bonferroni correction.</span>
<span class="co">#&gt; The p-value for feature 38 is 0.000559846.</span>
<span class="co">#&gt; This is significant at FWER = 0.05 after Bonferroni correction.</span></code></pre></div>
<p>We find six features that are statistically significant, which the user might erroneously report as discriminative features. We can visualize these six features by generating a Manhattan plot below.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Visualize differential analysis</span>
<span class="co"># Mutate the p-value vector for easier plotting</span>
<span class="va">p_val_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">p_val_vector</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">`colnames&lt;-`</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"p_value"</span>,<span class="st">"feature"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">select</span><span class="op">(</span><span class="va">feature</span>, <span class="va">p_value</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>log_p_value <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p_value</span><span class="op">)</span>,
         is_sig <span class="op">=</span> <span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p_value</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span> 

<span class="co"># Generate plot</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">p_val_vector</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">feature</span>, y <span class="op">=</span> <span class="va">log_p_value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>colour <span class="op">=</span> <span class="va">is_sig</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ylab</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span><span class="op">(</span><span class="op">-</span><span class="va">log</span><span class="op">[</span><span class="fl">10</span><span class="op">]</span><span class="op">(</span><span class="va">p</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">3</span>, lty <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_color_manual</span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#404040"</span>, <span class="st">"#d7191c"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"B"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,
        legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></code></pre></div>
<p><img src="intro_files/figure-html/2_5-1.png" width="700"></p>
<div id="fixing-the-problem" class="section level2">
<h2 class="hasAnchor">
<a href="#fixing-the-problem" class="anchor"></a>Fixing the Problem</h2>
<p>The process above, in which we first cluster the data and then perform differential analysis, is an example of <em>double dipping</em>.</p>
<p><details><summary><b>What is Double Dipping?</b></summary> Double dipping arises when insufficient care is taken in data analysis that involves multiple stages of computational work. While there are at least three well-documented instances of the phenomenon (<a href="https://www.nature.com/articles/nn.2303">Kriegerskorte et al., 2009</a>), here we focus on a two-stage pipeline involving clustering and differential analysis. We dip into the data first by clustering, and then dip again by performing differential analysis on the post-clustering data. This is problematic because the clustering method inherently searches for discriminative features to determine clusters. Thus we should not dip twice, or "double dip." </details></p>
<p>Fortunately, we can avoid risking double dipping by performing our exchangeability test on the dataset <span class="math inline">\(\mathbf{X}\)</span> as a first step.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## How to avoid risking double dipping</span>
<span class="co"># Compute p-value for test of homogeneity</span>
<span class="va">the_p_value</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPValue.html">getPValue</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"The p-value for the exchangeability test is "</span>, <span class="va">the_p_value</span>, <span class="st">".\n"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="co">#&gt; The p-value for the exchangeability test is 0.4868.</span>
<span class="kw">if</span> <span class="op">(</span><span class="va">the_p_value</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"No evidence of heterogeneity at alpha = 0.05!\n"</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">#&gt; No evidence of heterogeneity at alpha = 0.05!</span></code></pre></div>
<p>Note that alternative approaches, which we will not explore here, include selective inference (<a href="https://arxiv.org/abs/2012.02936">Gao et al., 2021+</a>). For this alternative approach, in the differential analysis step one computes a <span class="math inline">\(p\)</span>-value <em>conditioned on the cluster membership</em> obtained from the previous clustering step.</p>
</div>
</div>
<div id="dependent-features-example-from-population-genetics" class="section level1">
<h1 class="hasAnchor">
<a href="#dependent-features-example-from-population-genetics" class="anchor"></a>Dependent Features: Example from Population Genetics</h1>
<p>This section to be updated.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Alan Aw.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
