<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Population Stratification ‚Ä¢ flintyR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Population Stratification">
<meta property="og:description" content="flintyR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">flintyR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ex_vs_hom.html">Exchangeability and Homogeneity</a>
    </li>
    <li>
      <a href="../articles/extras.html">Extras</a>
    </li>
    <li>
      <a href="../articles/intro.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/ld_splitting.html">Linkage Disequilibrium Splitting</a>
    </li>
    <li>
      <a href="../articles/pop_strat.html">Population Stratification</a>
    </li>
    <li>
      <a href="../articles/single-cell-atac-seq.html">Analysis of scATAC-seq Data</a>
    </li>
    <li>
      <a href="../articles/wvs.html">Analysis of World Values Survey Data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/alanaw1/flintyR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Population Stratification</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/alanaw1/flintyR/blob/HEAD/vignettes/pop_strat.Rmd" class="external-link"><code>vignettes/pop_strat.Rmd</code></a></small>
      <div class="hidden name"><code>pop_strat.Rmd</code></div>

    </div>

    
    
<p><em>We are grateful to <a href="https://www.broadinstitute.org/bios/nick-patterson" class="external-link">Nick Patterson</a>, a Senior Computational Biologist in the Medical and Population Genetics Program at the Broad Institute of MIT and Harvard, for helpful discussions that spurred the creation of this vignette.</em></p>
<p>In our paper, we report how our methods can detect whether a set of genomes is exchangeable.</p>
<p>üìñ In this vignette, we will</p>
<ul>
<li>Compare our methods against <strong>EIGENSOFT</strong>, a stratification detection approach grounded in random matrix theory<br>
</li>
<li>Explore other approaches and potential extensions</li>
</ul>
<p>For pedagogical purposes, we code up all methods (except for <strong>flinty</strong>, of course) from scratch whenever possible. We welcome readers to re-use our code for their own work.</p>
<div class="section level2">
<h2 id="comparing-flinty-and-eigensoft">Comparing flinty and EIGENSOFT<a class="anchor" aria-label="anchor" href="#comparing-flinty-and-eigensoft"></a>
</h2>
<div class="section level3">
<h3 id="introduction-to-eigensoft-and-smartpca">Introduction to EIGENSOFT and smartpca<a class="anchor" aria-label="anchor" href="#introduction-to-eigensoft-and-smartpca"></a>
</h3>
<p><strong>EIGENSOFT</strong> (v7.2.1) is a software with multiple functionalities for genetic datasets: testing for stratification on genotype data, performing PCA on genotypes, and detecting causal variants whilst accounting for stratification. The first two functionalities are accomplished by the <strong>smartpca</strong> ‚Äúsubmodule‚Äù, whereas the latter is accomplished by another submodule named <strong>EIGENSTRAT</strong>. This Section focuses on <strong>smartpca</strong>.</p>
<p>As explained in our paper, an asymptotic result known as the Tracy-Widom theorem motivates a formal test of stratification relying on the largest eigenvalue of the sample covariance matrix. Concretely, suppose that <span class="math inline">\(\mathbf{X}\in\mathbb{R}^{N\times P}\)</span> consists of <span class="math inline">\(N\)</span> i.i.d. sub-Gaussian <span class="math inline">\(P\)</span>-vectors (e.g., each entry being iid <span class="math inline">\(N(0,1)\)</span> distributed satisfies this). Let <span class="math inline">\(s_1(\mathbf{X})\)</span> denote the largest singular value of <span class="math inline">\(\mathbf{X}\)</span>; this is equal to the square-root of the largest eigenvalue of <span class="math inline">\(\mathbf{X}\mathbf{X}^T\)</span>. Then, as <span class="math inline">\(N,P\rightarrow\infty\)</span> with <span class="math inline">\(P/N \rightarrow \gamma\in(0,\infty)\)</span>, Tracy-Widom theory says that <span class="math display">\[\begin{equation*}
\frac{s_1^2 - (\sqrt{N-1}+\sqrt{P})^2}{(\sqrt{N-1}+\sqrt{P})\left(\frac{1}{\sqrt{N-1}}+\frac{1}{\sqrt{P}}\right)^{1/3}} \overset{d}{\longrightarrow} F_1(s),
\end{equation*}\]</span><br>
where <span class="math inline">\(F_1\)</span> is the Tracy-Widom distribution with ensemble index <span class="math inline">\(1\)</span> (TW<span class="math inline">\({}_1\)</span>), i.e., <span class="math display">\[\begin{equation*}
F_1(x) = \exp\left(-\int_x^\infty [u(s)+(s-x)u^2(s)]ds\right)\hspace{1cm} \text{for }x \in \mathbb{R},
\end{equation*}\]</span> with <span class="math inline">\(u(s)\)</span> defined as the solution to the nonlinear ordinary differential equation <span class="math inline">\(u''=2u^3 + su\)</span> with asymptotic condition <span class="math inline">\(u(s)\sim \frac{1}{2\sqrt{\pi} s^{1/4}}\exp(-\frac{2}{3}s^{3/2})\)</span> as <span class="math inline">\(s\rightarrow\infty\)</span>. (The ODE is called the Painlev√© II equation and its solution the Hastings-Mcleod solution.)</p>
<p>Just like how the famous Central Limit Theorem allows us to construct confidence intervals and perform <span class="math inline">\(z\)</span>- or <span class="math inline">\(t\)</span>-tests on sample means that converge to the Gaussian under some null model, so does Tracy-Widom theory allow us to perform asymptotic tests on a multivariate sample, whose largest singular value converges to the TW<span class="math inline">\({}_1\)</span> under some null model. (Recall that the <span class="math inline">\(z\)</span>- and <span class="math inline">\(t\)</span>-tests are asymptotic two-sample tests of difference in sample means, contingent on the Central Limit Theorem working on the means even if the samples themselves are not normally distributed.)</p>
<p>What is the null model here? In population genetics, a panmictic population produces exchangeable individual genomes. In the presence of population structure, however, stratification of the sampled genomes occurs. Thus, in applying the TW test, we hope to detect stratification or exchangeability in a sample of genomes.</p>
<p>In a population-genetic setting, the rows of <span class="math inline">\(\mathbf{X}\)</span> are polarised genotypes or haplotypes, i.e., each row lies in either <span class="math inline">\(\{0,1,2\}^P\)</span> or <span class="math inline">\(\{0,1\}^P\)</span>. In <a href="https://doi.org/10.1371/journal.pgen.0020190" class="external-link">Patterson et al.¬†(2006)</a>, the authors propose the following test, in the absence of marker-marker linkages (i.e., LD).</p>
<ol style="list-style-type: decimal">
<li>
<p>Perform Hardy-Weinberg normalisation.</p>
<p>Let <span class="math inline">\(C(i,j)\)</span> be the number of variant alleles for marker <span class="math inline">\(j\)</span>, individual <span class="math inline">\(i\)</span>. Define the mean allelic dosage at marker <span class="math inline">\(j\)</span> by <span class="math inline">\(\mu(j) = \frac{\sum_{i=1}^N x_{ij}}{N}\)</span>, and define the estimate of the underlying allele frequency by <span class="math inline">\(p(j) = \mu(j)/2\)</span> if the data are of genotypes and <span class="math inline">\(p(j)=\mu(j)\)</span> if the data are of haplotypes. Compute the <em>Hardy-Weinberg normalised</em> version of <span class="math inline">\(\mathbf{X}\)</span>, <span class="math inline">\(\mathbf{M}=(m_{ij})\)</span>, where</p>
<p><span class="math display">\[\begin{equation*}
 m_{ij}=\frac{x_{ij} - \mu(j)}{\sqrt{p(j)(1-p(j))}}.
 \end{equation*}\]</span></p>
</li>
<li><p>Compute the sample covariance matrix, <span class="math inline">\(\mathbf{C} = \mathbf{M}\mathbf{M}^T\)</span>. This matrix has dimension <span class="math inline">\(N\times N\)</span>.</p></li>
<li><p>Order the eigenvalues of <span class="math inline">\(\mathbf{C}\)</span> so that <span class="math inline">\(\lambda_1&gt;\lambda_2&gt;\ldots&gt;\lambda_{N-1}&gt;0\)</span>. (Due to normalisation in Step 1, the last eigenvalue is zero. Hint: what is <span class="math inline">\(\mathbf{M}^T\mathbf{1}\)</span>?)</p></li>
<li>
<p>Compute the effective number of markers.</p>
<p>Instead of analysing the largest singular value of <span class="math inline">\(\mathbf{X}\)</span>, analyse the largest eigenvalue of <span class="math inline">\(\mathbf{C}\)</span>. This is because the TW<span class="math inline">\({}_1\)</span> approximation really applies to the <em>square</em> of the largest singular value, which is just the largest eigenvalue (i.e., <span class="math inline">\(s_1^2=\lambda_1\)</span>). Because <span class="math inline">\(\mathbf{C}\)</span> has an approximate Wishart distribution, one estimates the <a href="https://mathworld.wolfram.com/WishartDistribution.html" class="external-link">degrees of freedom parameter</a> of the Wishart, <span class="math inline">\(P'\)</span>, rather than using the putative number of features, <span class="math inline">\(P\)</span>:</p>
<p><span class="math display">\[\begin{equation*}
 P' = \frac{(N+1)\left(\sum_{i}\lambda_i\right)^2}{\left((N-1)\sum_{i}\lambda_i^2\right) - \left(\sum_{i}\lambda_i\right)^2}.
 \end{equation*}\]</span></p>
</li>
<li>
<p>Compute the normalised largest eigenvalue.</p>
<p>The normalised largest eigenvalue is defined by <span class="math display">\[\begin{equation*}
 \ell = \frac{(N-1) \lambda_1}{\sum_{i=1}^{N-1}\lambda_i}.
 \end{equation*}\]</span></p>
</li>
<li>
<p>Scale and centre <span class="math inline">\(\ell\)</span>.</p>
<p>Define the quantities <span class="math display">\[\begin{eqnarray*}
 a &amp; = &amp; \frac{(\sqrt{P'-1} + \sqrt{N})^2}{P'}, \\
 b &amp; = &amp; \frac{(\sqrt{P'-1} + \sqrt{N})}{P'}\left(\frac{1}{\sqrt{P'-1}} + \frac{1}{\sqrt{N}}\right)^{1/3},
 \end{eqnarray*}\]</span> where <span class="math inline">\(P'\)</span> is defined in Step 4. Compute <span class="math display">\[\begin{equation*}
 t = \frac{\ell - a}{b},
 \end{equation*}\]</span> where <span class="math inline">\(\ell\)</span> is defined in Step 5.</p>
</li>
<li>
<p>Compute the tail probability of <span class="math inline">\(t\)</span> against the TW<span class="math inline">\({}_1\)</span> distribution.</p>
<p>In other words, <span class="math inline">\(t\)</span> is approximately TW<span class="math inline">\({}_1\)</span>-distributed under an exchangeable, or ‚Äúno stratification‚Äù, null model.</p>
</li>
</ol>
<p>In <strong>smartpca</strong>, the <span class="math inline">\(p\)</span>-values reported in the output file are precisely the tail probabilities computed in Step 7.</p>
<p>Here is part of an example output file of <strong>smartpca</strong> that has to do with the test just described.</p>
<p style="text-align:center;">
<img src="smartpca_output.png" alt="EIGENSOFT smartpca example output." width="500"></p>
<p align="center">
<em>Example output file reporting significant eigenvalues of the sample covariance matrix, where p-values are computed as described above.</em>
</p>
<p>The reason there are multiple <span class="math inline">\(p\)</span>-values is that the software repeats the procedure above for all subsequent eigenvalues: if the top <span class="math inline">\(k\)</span> eigenvalues are declared significant, then the remaining eigenvalues <span class="math inline">\(\lambda_{k+1},\ldots,\lambda_{N}\)</span> can be tested for significance as though the normalised sample covariance matrix <span class="math inline">\(\mathbf{C}\)</span> (defined in Step 2) was a <span class="math inline">\((N-1 -k)\times (N-1-k)\)</span> Wishart matrix. (See ‚ÄúDetecting Additional Structure‚Äù on p.¬†2078 of Patterson et al., 2006).</p>
<details><summary><b>What does each column mean in the output file above?</b>
</summary>
The first column, <em>eigenvalue</em>, reports the <span class="math inline">\(k\)</span>th eigenvalue of the sample covariance matrix. The second column, <em>difference</em>, reports the difference between the <span class="math inline">\((k+1)\)</span>th and <span class="math inline">\(k\)</span>th eigenvalues in the previous column. The <em>twstat</em> column reports the scaled Tracy-Widom statistic, as described in Step 6 of the algorithm above. The <em>p-value</em> reports the tail probability, as described in Step 7. Finally, <em>effect.n</em> is the effective number of markers described in Step 4, which loosely speaking is a substitute estimate of the number of features (rather than using the number of features <span class="math inline">\(P\)</span>). See Step 4 for why this is done.
</details>
</div>
<div class="section level3">
<h3 id="independent-markers">Independent Markers<a class="anchor" aria-label="anchor" href="#independent-markers"></a>
</h3>
<p>In our paper we demonstrated that our method controls Type I Error. Can the same be said of the population-genetic <strong>EIGENSOFT</strong>? Since we believe that details matter to understanding <em>at the operational level</em>, we take care to elaborate on our simulation setup and the nuts and bolts of running the <strong>smartpca</strong> submodule. Subsequent sections will be presented with brevity.</p>
<p>We first consider how good <strong>EIGENSOFT/smartpca</strong> is at detecting exchangeability in the case of independent markers. We plan to examine performance on stratification in future work.</p>
<p>To simulate genomic data under an exchangeable null, we generate <span class="math inline">\(N=100\)</span> synthetic individual haplotypes, where each haplotype is comprised of <span class="math inline">\(P=200\)</span> independent markers, each realised by Bernoulli distributions with heads probability determined by a fixed population allele frequency ranging between <span class="math inline">\(0.2\)</span> and <span class="math inline">\(0.8\)</span>. This is shown in code below.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Helper function to draw samples</span>
<span class="co">## Function to generate N individual haplotypes </span>
<span class="co">## using simple coin flips, with heads probability </span>
<span class="co">## determined by the allele frequency. Allele frequencies</span>
<span class="co">## are arbitrarily drawn from a uniform distribution and fixed</span>
<span class="co">## beforehand. The second argument, num_ind_markers, decides </span>
<span class="co">## the number of independent alleles.</span>
<span class="va">getExHaplotypes</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">N</span>, <span class="va">num_ind_markers</span><span class="op">)</span> <span class="op">{</span>
  <span class="co"># Create population allele frequencies</span>
  <span class="va">allele_freqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">num_ind_markers</span>, min <span class="op">=</span> <span class="fl">0.2</span>, max <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span>
  
  <span class="co"># Generate array</span>
  <span class="va">out_array</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"cbind"</span>,
                       <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">allele_freqs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span><span class="va">N</span>,<span class="fl">1</span>,<span class="va">x</span><span class="op">)</span><span class="op">}</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">out_array</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"rs"</span>, <span class="fl">1</span><span class="op">:</span><span class="va">num_ind_markers</span><span class="op">)</span>
  
  <span class="co"># Return</span>
  <span class="va">to_return</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">out_array</span>,
                    bounds <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">num_ind_markers</span><span class="op">)</span>
  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">to_return</span><span class="op">)</span>
<span class="op">}</span>

<span class="co">## Generate array</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2021</span><span class="op">)</span>
<span class="va">ex_array</span> <span class="op">&lt;-</span> <span class="fu">getExHaplotypes</span><span class="op">(</span>N<span class="op">=</span><span class="fl">100</span>, num_ind_markers<span class="op">=</span><span class="fl">200</span><span class="op">)</span>

<span class="co">## Print</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">ex_array</span><span class="op">$</span><span class="va">data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;       rs1 rs2 rs3 rs4 rs5 rs6 rs7 rs8 rs9 rs10</span>
<span class="co">#&gt;  [1,]   0   1   1   1   0   0   1   1   1    1</span>
<span class="co">#&gt;  [2,]   0   1   0   0   0   0   1   0   1    1</span>
<span class="co">#&gt;  [3,]   1   1   1   0   0   1   0   0   1    0</span>
<span class="co">#&gt;  [4,]   1   1   1   1   1   0   1   1   0    1</span>
<span class="co">#&gt;  [5,]   0   1   1   0   0   0   0   0   1    1</span>
<span class="co">#&gt;  [6,]   0   1   1   0   1   0   0   0   1    1</span>
<span class="co">#&gt;  [7,]   0   0   1   0   0   1   1   0   1    0</span>
<span class="co">#&gt;  [8,]   0   1   1   0   0   1   1   0   0    1</span>
<span class="co">#&gt;  [9,]   0   1   1   1   0   1   0   0   1    1</span>
<span class="co">#&gt; [10,]   0   1   1   0   1   1   1   0   0    0</span></code></pre></div>
<p>To run <strong>smartpca</strong> on a simulated haplotype matrix, say <code>X_data</code>, we enter the following line in the directory containing the simulated matrix and the <strong>EIGENSOFT</strong> package. (We assume the user has these two files under the same directory.)</p>
<pre><code>./EIG-7.2.1/bin/smartpca -p X_data.pca.par &gt; EIGENSOFT_out</code></pre>
<p>Here <code>X_data.pca.par</code> is a colon-separated file providing information for <strong>smartpca</strong> to perform its computations:</p>
<p style="text-align:center;">
<img src="par_file.png" alt="EIGENSOFT par file." width="500"></p>
<p align="center">
<em>Example pca.par file to be read by EIGENSOFT/smartpca.</em>
</p>
<p>Additionally, with help from the <strong>RMTstat</strong> package (<a href="https://cran.rstudio.com/web/packages/RMTstat/index.html" class="external-link">Johnstone et al., 2014</a>), we implement from scratch the algorithm described in the beginning, just to see how well its performance agrees with the <strong>smartpca</strong> software. We use a two-tailed test for conservativeness.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Function to compute significance of largest eigenvalue </span>
<span class="co">## in the correlation matrix. We closely follow p. 2078 of</span>
<span class="co">## PPR2006.</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/evanbiederstedt/RMTstat" class="external-link">RMTstat</a></span><span class="op">)</span>
<span class="va">getTWPValue</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">M</span><span class="op">)</span> <span class="op">{</span>
  <span class="co"># 1. Assume M is computed as in Equations 1, 2, and 3.</span>
  <span class="co"># Use regressMatrix function to remove LD if necessary</span>
  <span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span>
  
  <span class="co"># 2. Compute X = MM'. X is m x m</span>
  <span class="va">X</span> <span class="op">&lt;-</span> <span class="va">M</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span>
  
  <span class="co"># 3. Order the eigenvalues of X so that ...</span>
  <span class="va">eigen_X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/eigen.html" class="external-link">eigen</a></span><span class="op">(</span><span class="va">X</span>, symmetric <span class="op">=</span> <span class="cn">TRUE</span>, only.values <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">$</span><span class="va">values</span>
  <span class="va">eigen_X</span> <span class="op">&lt;-</span> <span class="va">eigen_X</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="va">m</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">]</span>
  
  <span class="co"># 4. Estimate n' from Equation 10</span>
  <span class="va">n_prime</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">m</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">eigen_X</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span> <span class="op">/</span> <span class="op">(</span><span class="op">(</span><span class="op">(</span><span class="va">m</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">eigen_X</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">eigen_X</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span>
  
  <span class="co"># 5. The largest eigenvalue of M is lambda_1. Set l</span>
  <span class="va">lambda_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">eigen_X</span><span class="op">)</span>
  <span class="va">l</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">m</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="va">lambda_1</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">eigen_X</span><span class="op">)</span>
  
  <span class="co"># 6. Normalize l with Equations 5-7, where the effective number of markers n' replaces n</span>
  <span class="va">mu_mn</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">n_prime</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span> <span class="op">/</span> <span class="va">n_prime</span>
  <span class="va">sigma_mn</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">n_prime</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="va">n_prime</span> <span class="op">*</span> <span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">n_prime</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span>
  <span class="va">tw_stat</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">l</span> <span class="op">-</span> <span class="va">mu_mn</span><span class="op">)</span> <span class="op">/</span> <span class="va">sigma_mn</span>
  
  <span class="co"># Compute and return p-value</span>
  <span class="co"># Using two-tailed test for conservativeness</span>
  <span class="va">p_val</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/RMTstat/man/TracyWidom.html" class="external-link">ptw</a></span><span class="op">(</span><span class="va">tw_stat</span>, beta <span class="op">=</span> <span class="fl">1</span>, lower.tail <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/RMTstat/man/TracyWidom.html" class="external-link">ptw</a></span><span class="op">(</span><span class="va">tw_stat</span>, beta <span class="op">=</span> <span class="fl">1</span>, lower.tail <span class="op">=</span> <span class="cn">F</span><span class="op">)</span><span class="op">)</span>
  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">p_val</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>A functional script that performs the tasks above, and generates the results about to be reported, can be downloaded from <a href="https://github.com/alanaw1/exchangeability_test/blob/main/scripts/population_stratification/EIGENSOFT_EIGENSTRAT_demo.R" class="external-link">Github</a>. (You will need <strong>EIGENSOFT</strong> installed.)</p>
<p>We simulate <span class="math inline">\(200\)</span> exchangeable haplotypes matrices, with each matrix generated by <code>getExHaplotypes</code>. We estimate Type I Error rates by counting the fraction of false rejections of the null hypothesis at a specified significance threshold, <span class="math inline">\(\alpha\)</span>.</p>
<p>We find that <strong>smartpca</strong> controls Type I Error at commonly chosen significance thresholds (<span class="math inline">\(\alpha\in\{0.01,0.05\}\)</span>): even though the point estimates of Type I Error rate appear slightly above the nominal <span class="math inline">\(\alpha\)</span>, the <span class="math inline">\(95\%\)</span> confidence intervals well contain <span class="math inline">\(\alpha\)</span>.</p>
<p style="text-align:center;">
<img src="smartpca_FPR.jpg" alt="smartpca FPR" width="400"></p>
<p align="center">
<em>Type I Error Rates for smartpca.</em>
</p>
<p>We also find that our own implementation of the algorithm, which we refer to as <strong>R Implementation</strong>, is slightly conservative, albeit still controlling Type I Error (uniformly in <span class="math inline">\(\alpha\)</span>).</p>
<p style="text-align:center;">
<img src="R_implement_FPR.jpg" alt="smartpca FPR" width="400"></p>
<p align="center">
<em>Type I Error Rates for R Implementation. Note that an <a href="https://en.wikipedia.org/wiki/Exact_test" class="external-link">exact test</a> should have its empirical Type I error rate lying on or below the y=x line.</em>
</p>
<p>These results provide independent verification of the claim in Patterson et al.¬†(2006) that their method is efficacious for independent markers.</p>
</div>
<div class="section level3">
<h3 id="linked-markers">Linked Markers<a class="anchor" aria-label="anchor" href="#linked-markers"></a>
</h3>
<p>As we report briefly in our paper, we explored the performance of <strong>smartpca</strong> and <strong>flinty</strong> on simulated genomes with linked markers. To simulate such data, we rely on <strong>msprime</strong> to generate <span class="math inline">\(N=100\)</span> individual haplotypes across <span class="math inline">\(B=10\)</span> chromosomes, with each chromosome simulated independently of all other chromosomes. Per chromosome, haplotypes are generated from the coalescent with mutation rate and recombination rate set at <span class="math inline">\(2\times 10^{-8}\)</span> and effective population size <span class="math inline">\(10^4\)</span>. We further set the physical length of each chromosome to <span class="math inline">\(10^5\)</span>, so that each chromosome contains order <span class="math inline">\(10^2\)</span> single nucleotide polymorphisms. The code to generate SNPs within a single chromosome can be found <a href="https://github.com/alanaw1/exchangeability_test/blob/main/scripts/population_stratification/msprime_model.py" class="external-link">here</a>. Code to generate the <span class="math inline">\(10\)</span> chromosomes independently can be found in lines 78-106 of <a href="https://github.com/alanaw1/exchangeability_test/blob/main/scripts/population_stratification/EIGENSOFT_EIGENSTRAT_demo_dep.R" class="external-link">this script</a>.</p>
<p>Note that this generative model results in linked markers because the small recombination rates are insufficient to remove linkage.</p>
<p>On top of running <strong>smartpca</strong>, we perform LD pruning and marker regression, two strategies suggested by Patterson et al.¬†(2006) to remove linkage. We rely on the <strong>AssocTests</strong> software, recently developed by <a href="https://www.jstatsoft.org/article/view/v094i05" class="external-link">Wang et al.¬†(2020)</a> to run the <strong>smartpca</strong> algorithm within R, after performing LD pruning with <strong>PLINK</strong>, to facilitate smoother analysis. For marker regression, we code from scratch the method described on p.¬†2086 of Patterson et al.¬†(2006); see lines 5-33 in <a href="https://github.com/alanaw1/exchangeability_test/blob/main/scripts/population_stratification/EIGENSOFT_EIGENSTRAT_demo_dep.R" class="external-link">this script</a>. Consistent with the recommendations in the paper, we use <span class="math inline">\(k=5\)</span> markers in the regression, and pipe the regressed output into our previously coded up eigenvalue significance test, <code>getTWPValue</code>.</p>
<p>What do we find? Let us list the results by method. We begin by loading the results obtained from running <a href="https://github.com/alanaw1/exchangeability_test/blob/main/scripts/population_stratification/EIGENSOFT_EIGENSTRAT_demo_dep.R" class="external-link">our script</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load results</span>
<span class="co"># Load libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>

<span class="co"># Load result file</span>
<span class="va">EIGENSOFT_dep_df</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"EIGENSOFT_results_dep_df.csv"</span><span class="op">)</span></code></pre></div>
<ol style="list-style-type: decimal">
<li>
<strong>EIGENSOFT/smartpca</strong> (no correction for LD)</li>
</ol>
<p>The histogram below shows that at <span class="math inline">\(\alpha=0.05\)</span>, the number of significant eigenvalues is overestimated.</p>
<p style="text-align:center;">
<img src="EIGENSOFT_K_a005.jpg" alt="smartpca K plot" width="400"></p>
<p align="center">
<em>Histogram of the number of significant eigenvalues, K. For an exchangeable sequence, K should be 0.</em>
</p>
<p>Because <span class="math inline">\(K=0\)</span> is the typical decision rule for whether a sample is unstratified (i.e., exchangeable), we can compute the FPR at <span class="math inline">\(\alpha=0.05\)</span>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Compute EIGENSOFT FPR</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"The FPR estimate is "</span>, <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">EIGENSOFT_dep_df</span><span class="op">$</span><span class="va">EIGENSOFT_a5</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>, <span class="st">"."</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] "The FPR estimate is 1."</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Our R Implementation (<code>getTWPValue</code>), with marker regression to correct for LD</li>
</ol>
<p>The plot of Type I Error against significance threshold <span class="math inline">\(\alpha\)</span> shows that the conservative test of non-exchangeability also has inflated Type I error.</p>
<p style="text-align:center;">
<img src="R_implement_dep_FPR.jpg" alt="R Implementation FPR" width="400"></p>
<p align="center">
<em>Type I Error Rates for R Implementation, with simulated markers in LD.</em>
</p>
<p>This demonstrates that marker regression alone can be inadequate at accounting for LD whilst deciding whether a sample of genomes is exchangeable.</p>
<ol start="3" style="list-style-type: decimal">
<li>
<strong>AssocTests</strong>, with LD pruning</li>
</ol>
<p>LD pruning was performed using PLINK, such that whenever a pair of variants with LD at least <span class="math inline">\(0.1\)</span> within a window of <span class="math inline">\(50\)</span> SNPs is observed, the variant with lower MAF is removed. (You can see line 165 of <a href="https://github.com/alanaw1/exchangeability_test/blob/main/scripts/population_stratification/EIGENSOFT_EIGENSTRAT_demo_dep.R" class="external-link">our script</a> and the <a href="https://groups.google.com/g/plink2-users/c/FSAF_36tIEs" class="external-link">PLINK2 users forum</a> for details.)</p>
<p>The plot below shows that at <span class="math inline">\(\alpha=0.05\)</span>, the number of significant eigenvalues is overestimated.</p>
<p style="text-align:center;">
<img src="EIGENSTRAT_K_a005.jpg" alt="smartpca K plot" width="400"></p>
<p align="center">
<em>Histogram of the number of significant eigenvalues, K. For an exchangeable sequence, K should be 0.</em>
</p>
<p>This demonstrates that LD pruning alone is also inadequate at accounting for LD whilst deciding whether a sample of genomes is exchangeable.</p>
<ol start="4" style="list-style-type: decimal">
<li><strong>flinty</strong></li>
</ol>
<p>We assume the variants are partitionable, and moreover assume that variants lying in different chromosomes are independent of one another. (This assumption, if true, enables us to build chromosome-specific evolutionary models with the individual haplotypes as data.)</p>
<p>The plot of Type I Error against significance threshold <span class="math inline">\(\alpha\)</span> shows that <strong>flinty</strong> reasonably controls the Type I error. Even though the point estimates sometimes lie above the <span class="math inline">\(y=x\)</span> line, the <span class="math inline">\(95\%\)</span> confidence intervals well contain the <span class="math inline">\(y=x\)</span> curve.</p>
<p style="text-align:center;">
<img src="flinty_dep_FPR.jpg" alt="smartpca K plot" width="400"></p>
<p align="center">
<em>Type I Error Rates for flinty, with simulated markers in LD.</em>
</p>
</div>
</div>
<div class="section level2">
<h2 id="other-approaches">Other Approaches<a class="anchor" aria-label="anchor" href="#other-approaches"></a>
</h2>
<div class="section level3">
<h3 id="alternative-normalisation-strategies-for-tw-statistic">Alternative Normalisation Strategies for TW Statistic<a class="anchor" aria-label="anchor" href="#alternative-normalisation-strategies-for-tw-statistic"></a>
</h3>
<p>This section will explore different approaches used to estimate the ‚Äúeffective number of markers‚Äù parameter of the observed Wishart matrix computed from the scaled genotype matrix. To be continued.</p>
</div>
<div class="section level3">
<h3 id="empirical-spectral-distribution">Empirical Spectral Distribution<a class="anchor" aria-label="anchor" href="#empirical-spectral-distribution"></a>
</h3>
<p>This section will explore other spectral statistics potentially useful for detecting structure. To be continued.</p>
</div>
<div class="section level3">
<h3 id="parallel-analysis">Parallel Analysis<a class="anchor" aria-label="anchor" href="#parallel-analysis"></a>
</h3>
<p>To be continued.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alan Aw.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
