<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of scATAC-seq Data • flintyR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of scATAC-seq Data">
<meta property="og:description" content="flintyR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">flintyR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ex_vs_hom.html">Exchangeability and Homogeneity</a>
    </li>
    <li>
      <a href="../articles/intro.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/single-cell-atac-seq.html">Analysis of scATAC-seq Data</a>
    </li>
    <li>
      <a href="../articles/wvs.html">Analysis of World Values Survey Data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="single-cell-atac-seq_files/header-attrs-2.7/header-attrs.js"></script><script src="single-cell-atac-seq_files/kePrint-0.0.1/kePrint.js"></script><link href="single-cell-atac-seq_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysis of scATAC-seq Data</h1>
            
      
      
      <div class="hidden name"><code>single-cell-atac-seq.Rmd</code></div>

    </div>

    
    
<p>In cellular biology, the study of single cells allows biologists to discover mechanisms not detectable by studying only bulk populations of cells. In recent years, advances in sequencing technologies have enabled the quantification of many single cell features (e.g., transcriptome, proteome, metabolome, 3D contact maps), and such datasets have in turn provided new statistical challenges and spurred the development of new probabilistic models (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-1926-6">Lähnemann et al., 2020</a>).</p>
<p>Key to discovering mechanisms from single cell data are analyses of heterogeneity between single cell populations. Heterogeneous patterns across multiple biological features reveal biological differences between groups of cells that differ by physical location, temporal factors, or some other biological function, and such differences can subsequently help generate hypotheses relating features to function. Moreover, depending on the question the resolution of the population might differ, for instance at the tissue level, at the cell type level, or even at physical location level among the same cell types.</p>
<p style="text-align:center;">
<img src="ex_data/scATAC-seq/bmc_fig1.png" alt="Many resolutions of single cell statistical inference." width="500"></p>
<p align="center">
<em>Many levels of resolution at which to investigate single cells. Figure reproduced from <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-1926-6">Lähnemann et al. (2020)</a> and <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1663-x">Wolf et al. (2019)</a> under the Creative Commons Attribution (CC-BY) license.</em>
</p>
<p>Central to single cell analyses is the clustering of a large population of sampled cells from an experiment. Such clustering often partitions the large population into single cell communities that correspond to some known biological function (e.g., cell type or sampling batch). Once these cell communities are obtained, further analyses can be conducted <em>on</em> or <em>between</em> identified communities to generate biological hypotheses.</p>
<p>📖 In this vignette, we will:</p>
<ul>
<li>show how <strong>flinty</strong> can be applied to detect statistical heterogeneity in single cell clusters detected from scATAC-seq data,</li>
<li>demonstrate how <strong>flinty</strong> can explain drivers of the detected heterogeneity.</li>
</ul>
<div id="single-cell-atac-seq-data" class="section level1">
<h1 class="hasAnchor">
<a href="#single-cell-atac-seq-data" class="anchor"></a>Single Cell ATAC-seq Data</h1>
<p>We obtain publicly available human peripheral blood mononuclear cell (PBMC) scATAC-seq data from <a href="https://www.10xgenomics.com/">10X Genomics</a>. Data processing requires <strong>Seurat</strong> and its extension, <strong>Signac (v0.2.5)</strong>; both can be installed with Bioconductor.</p>
<p>The code below loads and describes the PBMC data <em>before</em> running data processing steps.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/timoast/signac">Signac</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomeInfoDb">GenomeInfoDb</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://hhoeflin.github.io/hdf5r">hdf5r</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span>

<span class="co">## Load the original data</span>
<span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X_h5.html">Read10X_h5</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="st">"ex_data/scATAC-seq/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5"</span><span class="op">)</span>
<span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span>
  file <span class="op">=</span> <span class="st">"ex_data/scATAC-seq/atac_v1_pbmc_10k_singlecell.csv"</span>,
  header <span class="op">=</span> <span class="cn">TRUE</span>,
  row.names <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span>

<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>
  counts <span class="op">=</span> <span class="va">counts</span>,
  assay <span class="op">=</span> <span class="st">'peaks'</span>,
  project <span class="op">=</span> <span class="st">'ATAC'</span>,
  min.cells <span class="op">=</span> <span class="fl">1</span>,
  meta.data <span class="op">=</span> <span class="va">metadata</span>
<span class="op">)</span>

<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/signac/reference/SetFragments.html">SetFragments</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">pbmc</span>,
  file <span class="op">=</span> <span class="st">"ex_data/scATAC-seq/atac_v1_pbmc_10k_fragments.tsv.gz"</span>
<span class="op">)</span> 

<span class="co">## Show the PBMC object before data processing</span>
<span class="va">pbmc</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 79921 features across 9277 samples within 1 assay </span>
<span class="co">#&gt; Active assay: peaks (79921 features, 0 variable features)</span></code></pre></div>
<div id="data-processing" class="section level2">
<h2 class="hasAnchor">
<a href="#data-processing" class="anchor"></a>Data Processing</h2>
<p>We perform all steps of data processing up to <a href="https://satijalab.org/signac/articles/pbmc_vignette.html#non-linear-dimension-reduction-and-clustering-1">non-linear dimension reduction and clustering</a> in this <a href="https://satijalab.org/signac/articles/pbmc_vignette.html">tutorial</a> provided by the lab of Rahul Satija.</p>
<p>To summarize the essential steps, we</p>
<ol style="list-style-type: decimal">
<li>Filter individual cells that fail to satisfy criteria associated with five quality control metrics (nucleosome banding pattern, TSS enrichment score, total number of fragments in peaks, fraction of fragments in peaks and ratio reads in blacklist sites provided by ENCODE);</li>
<li>Perform term frequency-inverse document frequency (TF-IDF) normalization and singular value decomposition (SVD) on the TF-IDF normalized matrix;</li>
<li>Cluster the cells using the smart local moving (SLM) algorithm.</li>
</ol>
<p>We obtain <span class="math inline">\(6954\)</span> cells after Step 1, and Step 3 yielded <span class="math inline">\(15\)</span> clusters altogether, whose sizes are reported below. Note Step 2 is performed to facilitate Step 3, and we still keep the un-normalized data from Step 1 to run our statistical tests.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load summary of clustering, show cluster sizes</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"ex_data/scATAC-seq/scATAC-seq_annotation_cluster.RData"</span><span class="op">)</span>
<span class="fu">kableExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">satija.cell.annot.cluster</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span>,
                  col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Cluster"</span>, <span class="st">"Count"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">kableExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/kableExtra/man/kable_styling.html">kable_styling</a></span><span class="op">(</span>position <span class="op">=</span> <span class="st">"center"</span><span class="op">)</span></code></pre></div>
<table class="table table">
<thead><tr>
<th style="text-align:left;">
Cluster
</th>
<th style="text-align:right;">
Count
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
0
</td>
<td style="text-align:right;">
2494
</td>
</tr>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:right;">
732
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:right;">
627
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:right;">
569
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:right;">
563
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:right;">
432
</td>
</tr>
<tr>
<td style="text-align:left;">
6
</td>
<td style="text-align:right;">
348
</td>
</tr>
<tr>
<td style="text-align:left;">
7
</td>
<td style="text-align:right;">
277
</td>
</tr>
<tr>
<td style="text-align:left;">
8
</td>
<td style="text-align:right;">
245
</td>
</tr>
<tr>
<td style="text-align:left;">
9
</td>
<td style="text-align:right;">
216
</td>
</tr>
<tr>
<td style="text-align:left;">
10
</td>
<td style="text-align:right;">
199
</td>
</tr>
<tr>
<td style="text-align:left;">
11
</td>
<td style="text-align:right;">
122
</td>
</tr>
<tr>
<td style="text-align:left;">
12
</td>
<td style="text-align:right;">
62
</td>
</tr>
<tr>
<td style="text-align:left;">
13
</td>
<td style="text-align:right;">
50
</td>
</tr>
<tr>
<td style="text-align:left;">
14
</td>
<td style="text-align:right;">
18
</td>
</tr>
</tbody>
</table>
<p>These clusters can be visualized in 2D space using UMAP. The code to reproduce this plot is provided in the Satija lab vignettes linked above.</p>
<p style="text-align:center;">
<img src="ex_data/scATAC-seq/pbmc_umap_plot.jpg" alt="UMAP showing separation of clusters." width="400"></p>
<p>Furthermore, these clusters correspond to myeloid and lymphoid cell populations. One can identify cell types by identifying cluster-defining biological markers using RNA-seq measurements of the same cells; see the Satija lab vignettes <a href="https://satijalab.org/seurat/v3.0/pbmc3k_tutorial.html">here</a> and <a href="https://satijalab.org/seurat/v3.1/atacseq_integration_vignette.html">there</a> for details on accomplishing this “feature anchoring” step.</p>
<p style="text-align:center;">
<img src="ex_data/scATAC-seq/cell_annotation.jpg" alt="Pie charts summarizing cell type composition per cluster" width="800"></p>
<p>Before proceeding with our analysis, we note that the <span class="math inline">\(79921\)</span> features per individual cell correspond to physical regions along the human genome, and each feature records the number of reads, more precisely Tn5cut sites, mapped by that individual cell to the physical region. These read-based records measure chromatin activity and thus provide information about gene regulation.</p>
</div>
</div>
<div id="analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis" class="anchor"></a>Analysis</h1>
<p>Data processing has provided us with a dataset <span class="math inline">\(\mathbf{X}\in\mathbb{N}^{N\times P}\)</span> consisting of <span class="math inline">\(N=6954\)</span> single cells with <span class="math inline">\(P=79921\)</span> peak reads/counts features recorded per cell. Additionally, <span class="math inline">\(\mathbf{X}\)</span> has been partitioned into <span class="math inline">\(15\)</span> clusters that correspond to certain distinguishable blood cell populations.</p>
<p>Our first goal is to check whether each cluster of cells looks homogeneous. Indeed, there may be latent stratification or heterogeneities unaccounted for during clustering.</p>
<p>We first binarize <span class="math inline">\(\mathbf{X}=(x_{ij})\)</span> by setting each entry <span class="math inline">\(x_{ij}=1\)</span> if it is positive, and <span class="math inline">\(x_{ij}=0\)</span> otherwise. Such data compression results in loss of information. However, owing to our compression step itself <em>reducing</em> variability, if our test reports non-exchangeability on the compressed version, then, all other things being equal, it is the case that the original sample exhibits heterogeneity. (This is a general intuition that may not be true if there exist appropriate statistical models for the original features in the uncompressed version. Hence, we place the caveat “all things being equal.” To our knowledge, such a model still does not exist.)</p>
<p>Here is a snapshot of the binarized version of samples from the smallest cluster (Cluster 14).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Show Cluster 15 (first six cells and first six features) after binarization</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"ex_data/scATAC-seq/bin_satija_keepzeros_15.RData"</span><span class="op">)</span>
<span class="va">subsetted_data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>
<span class="co">#&gt; 6 x 6 sparse Matrix of class "dgCMatrix"</span>
<span class="co">#&gt;                    chr1:565113-565543 chr1:569179-569635 chr1:713534-714806</span>
<span class="co">#&gt; AAACTGCGTGCTGAAG-1                  .                  .                  .</span>
<span class="co">#&gt; AATGGAATCCCACGGA-1                  .                  .                  1</span>
<span class="co">#&gt; ACAGCGCAGCCCATTA-1                  .                  .                  .</span>
<span class="co">#&gt; ACCATCCCACGCGACT-1                  .                  .                  1</span>
<span class="co">#&gt; ACGGATTTCTTCCGTG-1                  .                  .                  .</span>
<span class="co">#&gt; ACTCGGGTCCCGCAAG-1                  .                  .                  1</span>
<span class="co">#&gt;                    chr1:752436-753020 chr1:762144-763353 chr1:779610-780252</span>
<span class="co">#&gt; AAACTGCGTGCTGAAG-1                  .                  1                  .</span>
<span class="co">#&gt; AATGGAATCCCACGGA-1                  .                  .                  .</span>
<span class="co">#&gt; ACAGCGCAGCCCATTA-1                  .                  .                  .</span>
<span class="co">#&gt; ACCATCCCACGCGACT-1                  .                  .                  .</span>
<span class="co">#&gt; ACGGATTTCTTCCGTG-1                  .                  .                  .</span>
<span class="co">#&gt; ACTCGGGTCCCGCAAG-1                  .                  .                  .</span></code></pre></div>
<p>Next, owing to the large number of peak regions (<span class="math inline">\(P=79921\)</span>) in our scATAC-seq data, for our analysis below we restrict our features to only those peak regions that lie in Chromosome 1. This restriction reduced the number of features to <span class="math inline">\(P=7824\)</span>. Again, we show an example for cells in Cluster 14.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Restricting to Chromosome 1 only</span>
<span class="co"># Get dimension of the array before restriction to peak regions lying in chr1</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">subsetted_data</span><span class="op">)</span>
<span class="co">#&gt; [1]    18 79921</span>
<span class="co"># Restrict peak regions to lie in chr1</span>
<span class="va">subsubsetted_data</span> <span class="op">&lt;-</span> <span class="va">subsetted_data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"chr1:"</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">subsetted_data</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>
<span class="co"># Get dimension of subarray containing only selected features</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">subsubsetted_data</span><span class="op">)</span>
<span class="co">#&gt; [1]   18 7824</span></code></pre></div>
<div id="contingency-of-test" class="section level2">
<h2 class="hasAnchor">
<a href="#contingency-of-test" class="anchor"></a>Contingency of Test</h2>
<p>One important handle of our test is the specification of dependencies between features. It is unlikely that the features, which are peak reads, are mutually independent of one another, since peak regions within a larger region of open chromatin may have proximal genes that are co-regulated, which drive peak reads themselves to be statistically correlated. Thus we have to account for some dependency structure.</p>
<p>Here, one advantage is that the user has control over what kinds of dependencies to account for, which makes the test interpretable. For example, accounting for dependencies arising from 3D organization of the genome would mean that the test is performed and its result valid, contingent exactly on such 3D organizational dependencies being the only source of dependence between features.</p>
<p>We will in fact account for dependencies arising from 3D organization of the genome. There has been speculation that the 3D structure of the genome is the primary driver of differences in chromatin accessibility. Because chromatin 3D organization is measured by probing physical interaction between non-adjacent genomic loci and such probing in turn is accomplished by chromatin conformation capture (3C) technologies (<a href="https://science.sciencemag.org/content/295/5558/1306">Dekker et al., 2002</a>), here we shall incorporate Hi-C information (<a href="https://science.sciencemag.org/content/326/5950/289">Lieberman-Aiden et al., 2009</a>) into our tests.</p>
<p>Concretely, we use Hi-C data to partition our features into disjoint sets of dependent features (so-called “dependent blocks”), where we group features into the same sets based on whether they belong in the same topologically associating domain (TAD) as identified by Hi-C TAD callers.</p>
</div>
<div id="methods-to-call-tads-on-hi-c-data" class="section level2">
<h2 class="hasAnchor">
<a href="#methods-to-call-tads-on-hi-c-data" class="anchor"></a>Methods to Call TADs on Hi-C Data</h2>
<p>There are more than 20 TAD calling approaches; see <a href="https://link.springer.com/article/10.1007/s12551-018-0489-1">Pal et al. (2019)</a> for a summary of them. We consider two approaches, <strong>DomainCaller</strong> — a method proposed by <a href="https://www.nature.com/articles/nature11082">Dixon et al. (2012)</a> that relies on the directionality index — and <strong>TopDom</strong> — a newer approach that scores patterns of contact frequencies and tests these scores for statistical significance under a non-parametric null distribution <a href="https://academic.oup.com/nar/article/44/7/e70/2467818">Shin et al. (2016)</a>. Note both <strong>DomainCaller</strong> and <strong>TopDom</strong> are found reliable in a comprehensive meta-analysis of 22 TAD callers by <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1596-9">Zufferey et al. (2018)</a>.</p>
<details><summary><b>A Important Caveat</b>
</summary>
Because the single cell data provided by 10X Genomics is not accompanied by Hi-C contact maps of the same cells, we instead rely on an external Hi-C dataset consisting of human fetal lung fibroblast (IMR90) cells (<a href="https://www.nature.com/articles/nature11082">Dixon et al., 2012</a>). This introduces one caveat to our analysis: we assume TADs identified from IMR90 cells are approximately identical to TADs in PBMC. Note that there is some evidence suggesting that TADs are reasonably preserved across cell types; see <a href="https://academic.oup.com/bioinformatics/article/34/13/i475/5045717">Sauerwald and Kingsford (2018)</a> for instance. Careful quantification of the invariance of TADs toward cell type specification is a topic of ongoing research, and we direct the interested reader to recent work, e.g., <a href="https://www.nature.com/articles/s41467-018-03017-1">Gong et al. (2018)</a>, <a href="https://bmcbiol.biomedcentral.com/articles/10.1186/s12915-018-0556-x">Krefting et al. (2018)</a>, and <a href="https://www.sciencedirect.com/science/article/abs/pii/S000292972100001X">McArthur and Capra (2021)</a>.
</details>
</div>
</div>
<div id="results" class="section level1">
<h1 class="hasAnchor">
<a href="#results" class="anchor"></a>Results</h1>
<div id="explanations" class="section level2">
<h2 class="hasAnchor">
<a href="#explanations" class="anchor"></a>Explanations</h2>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Alan Aw.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
