<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction • flintyR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction">
<meta property="og:description" content="flintyR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">flintyR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ex_vs_hom.html">Exchangeability and Homogeneity</a>
    </li>
    <li>
      <a href="../articles/extras.html">Extras</a>
    </li>
    <li>
      <a href="../articles/intro.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/single-cell-atac-seq.html">Analysis of scATAC-seq Data</a>
    </li>
    <li>
      <a href="../articles/wvs.html">Analysis of World Values Survey Data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/alanaw1/flintyR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="intro_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/alanaw1/flintyR/blob/master/vignettes/intro.Rmd"><code>vignettes/intro.Rmd</code></a></small>
      <div class="hidden name"><code>intro.Rmd</code></div>

    </div>

    
    
<strong>flinty</strong> provides scientists a way to test if their datasets contain multiple subpopulations. This is an important question if the scientist wishes to infer quantities of interest from the dataset, for instance population-level parameters, without overlooking any latent heterogeneities.
<p align="center">
<img src="intro_visual.png" width="600"></p>
<p align="center">
<em>Bruno sees the array in front of him but does not see the latent information at the back.</em>
</p>
<p>Consider the following examples commonly encountered by practitioners.</p>
<ul>
<li>In population genetics, genetic samples are collected from individuals, say in a country like Mexico whose population history records admixture events and endogamous mating by cultural ancestry. It is not surprising if a dataset contains individuals that belong to distinguishable genetic communities. Unfortunately, absent information like family pedigrees, we might not have external means of grouping the individuals. This leads to less fine-scale analysis of the data and missed inferences about subpopulations.</li>
<li>In single cell genomics, unique molecular identifiers (UMIs) are collected for a large sample of cells, and clustering algorithms are run to generate homogeneous groups of cells. External validation is used to guide correctness of clustering, and the resulting groups are assumed homogeneous for downstream analysis. However, the clustering algorithm does not <em>provably</em> recover homogeneous clusters without mathematical justification and underlying modeling assumptions made about the data. This has prompted the use of <a href="https://doi.org/10.1080/01621459.2020.1833888">statistical hypothesis testing</a> to validate post-clustering homogeneity.<br>
</li>
<li>In survey questionnaires typical of the social sciences, the multivariate dataset <span class="math inline">\(\mathbf{X}\)</span> conceals latent factors that can be recovered statistically. Concretely, it is assumed that the <span class="math inline">\(P\)</span>-variate samples <span class="math inline">\(\mathbf{x}_1,\ldots,\mathbf{x}_N\)</span> making up <span class="math inline">\(\mathbf{X}\)</span> satisfy <span class="math display">\[\mathbf{x}_i = \mathbf{\Lambda}\boldsymbol{\xi}_i + \boldsymbol{\delta}_i\]</span> for each sample <span class="math inline">\(i\)</span>. Here <span class="math inline">\(\mathbf{\Lambda}\)</span> is the <span class="math inline">\(P\times M\)</span> matrix of loadings, <span class="math inline">\(\boldsymbol{\xi}_i\)</span> is the <span class="math inline">\(M\)</span>-dimensional sample-specific measurement at those <span class="math inline">\(M\)</span> factors, and <span class="math inline">\(\boldsymbol{\delta}_i\)</span> is a <span class="math inline">\(P\)</span>-dimensional residual vector. After learning the latent factors from the dataset (i.e., <span class="math inline">\(\mathbf{\Lambda}\)</span>, and by extension <span class="math inline">\(\boldsymbol{\xi}_i\)</span> and <span class="math inline">\(\boldsymbol{\delta}_i\)</span> for each <span class="math inline">\(i\)</span>), it is important to <a href="https://doi.org/10.1207/S15327906MBR3801_5">check that</a> the residual vectors <span class="math inline">\(\mathbf{\Delta}=[\boldsymbol{\delta}_1,\ldots,\boldsymbol{\delta}_N]^T\)</span> look independent and identically distributed, or at least exchangeable, so that no other latent factor was missed.</li>
</ul>
<p>For each example above, methods have been developed to address the issue described. These methods vary in their degree of success. We won’t discuss the successes and limitations of such methods here. We merely want to suggest connections of our methods with some of these settings.</p>
<p>Instead, in this vignette we shall describe our general method of exchangeability and demonstrate its use on simulated data. Along the way, we will delight ourselves with a few morsels of data analysis, including <em>double dipping</em> and <em>runs tests</em>.</p>
<div id="setup" class="section level1">
<h1 class="hasAnchor">
<a href="#setup" class="anchor"></a>Setup</h1>
<p>We start by loading <strong>flintyR</strong> alongside other useful packages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://alanaw1.github.io/flintyR">flintyR</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">doParallel</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/doParallel/man/registerDoParallel.html">registerDoParallel</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div id="basic-example" class="section level1">
<h1 class="hasAnchor">
<a href="#basic-example" class="anchor"></a>Basic Example</h1>
<p>We start simple. This example is motivated by recent work of Lucy Gao and Daniela Witten, which they presented in the International Seminar on Selective Inference and in the R/Medicine 2020 Virtual Conference respectively.</p>
<p>We simulate <span class="math inline">\(\mathbf{X}\)</span>, a dataset consisting of <span class="math inline">\(N=100\)</span> samples, where each sample consists of <span class="math inline">\(P=50\)</span> independent features, each feature a realization of a standard Gaussian random variable. Mathematically, the entries <span class="math inline">\(x_{np}\overset{\text{iid}}{\sim} N(0,1)\)</span>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Generate some data</span>
<span class="co"># Set seed for reproducibility</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">420</span><span class="op">)</span>
<span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">5000</span><span class="op">)</span>,
            nrow <span class="op">=</span> <span class="fl">100</span>,
            ncol <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></code></pre></div>
<p>Under such settings <span class="math inline">\(\mathbf{X}\)</span> is homogeneous and the samples <span class="math inline">\(\mathbf{x}_1,\ldots,\mathbf{x}_N\)</span> are exchangeable, but if we were handed <span class="math inline">\(\mathbf{X}\)</span> without knowing its generative mechanism completely, we might want to perform clustering as part of exploratory data analysis.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Cluster data</span>
<span class="va">X_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span>
<span class="va">X_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span><span class="op">(</span><span class="va">X_df</span>, centers <span class="op">=</span> <span class="fl">2</span>, nstart <span class="op">=</span> <span class="fl">25</span><span class="op">)</span>

<span class="co"># Assign cluster labels</span>
<span class="va">X_df</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="va">X_cluster</span><span class="op">$</span><span class="va">cluster</span>

<span class="co"># Perform PCA to visualize</span>
<span class="va">X_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/princomp.html">princomp</a></span><span class="op">(</span><span class="va">X_df</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span><span class="op">)</span>
<span class="va">PC1and2</span> <span class="op">&lt;-</span> <span class="va">X_pca</span><span class="op">$</span><span class="va">scores</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">PC1and2</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="va">X_cluster</span><span class="op">$</span><span class="va">cluster</span> </code></pre></div>
<p>By performing <span class="math inline">\(k\)</span>-means clustering with <span class="math inline">\(k=2\)</span> as shown above, we obtain a cluster assignment. We can visualize this clustering by performing PCA on the samples and plotting the projection of each sample onto its first two principal components.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Visualize clustering</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">PC1and2</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Comp.1</span>, y <span class="op">=</span> <span class="va">Comp.2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span>,
             shape <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">xlab</span><span class="op">(</span><span class="st">"PC1"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ylab</span><span class="op">(</span><span class="st">"PC2"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>colour<span class="op">=</span><span class="fu">guide_legend</span><span class="op">(</span>title<span class="op">=</span><span class="st">"Cluster"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"A"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="figure/2_3-1.png" alt=""><p class="caption">plot of chunk 2_3</p>
</div>
<p>Next, we perform <a href="https://en.wikipedia.org/wiki/Mann%E2%80%93Whitney_U_test">Wilcoxon rank sum tests</a> between the two populations for each feature, applying Bonferroni correction to control the family-wise error rate (FWER) at <span class="math inline">\(\alpha=0.05\)</span>. Recall that the Wilcoxon rank sum (aka, Mann-Whitney) test is a non-parametric two-sample test for difference in sample distributions.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Differential analysis</span>
<span class="co"># Split into two clusters </span>
<span class="va">X_df_1</span> <span class="op">&lt;-</span> <span class="va">X_df</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> 
<span class="co">#&gt; Error in filter(., cluster == 1): object 'cluster' not found</span>
<span class="va">X_df_2</span> <span class="op">&lt;-</span> <span class="va">X_df</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span> 
<span class="co">#&gt; Error in filter(., cluster == 2): object 'cluster' not found</span>

<span class="co"># Compute p-values</span>
<span class="va">p_val_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span> <span class="op">{</span>
  <span class="co"># Perform differential analysis on feature j</span>
  <span class="va">da_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/wilcox.test.html">wilcox.test</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">X_df_1</span><span class="op">[</span>,<span class="va">j</span><span class="op">]</span>, 
                           y <span class="op">=</span> <span class="va">X_df_2</span><span class="op">[</span>,<span class="va">j</span><span class="op">]</span><span class="op">)</span>
  <span class="co"># Print result if significant</span>
  <span class="va">p_val_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">p_val_vector</span>, <span class="va">da_result</span><span class="op">$</span><span class="va">p.value</span><span class="op">)</span>
  <span class="kw">if</span> <span class="op">(</span><span class="va">da_result</span><span class="op">$</span><span class="va">p.value</span> <span class="op">&lt;</span> <span class="fl">0.001</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"The p-value for feature "</span>, <span class="va">j</span>, <span class="st">" is "</span>, <span class="va">da_result</span><span class="op">$</span><span class="va">p.value</span>, <span class="st">".\n"</span>, 
        sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"This is significant at FWER = 0.05 after Bonferroni correction.\n"</span><span class="op">)</span>
  <span class="op">}</span>
<span class="op">}</span>
<span class="co">#&gt; Error in wilcox.test(x = X_df_1[, j], y = X_df_2[, j]): object 'X_df_1' not found</span></code></pre></div>
<p>We find six features that are statistically significant, which the user might erroneously report as discriminative features. We can visualize these six features by generating a Manhattan plot below.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Visualize differential analysis</span>
<span class="co"># Mutate the p-value vector for easier plotting</span>
<span class="va">p_val_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">p_val_vector</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">`colnames&lt;-`</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"p_value"</span>,<span class="st">"feature"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">select</span><span class="op">(</span><span class="va">feature</span>, <span class="va">p_value</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>log_p_value <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p_value</span><span class="op">)</span>,
         is_sig <span class="op">=</span> <span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p_value</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span> 
<span class="co">#&gt; Error in names(x) &lt;- value: 'names' attribute [2] must be the same length as the vector [1]</span>

<span class="co"># Generate plot</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">p_val_vector</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">feature</span>, y <span class="op">=</span> <span class="va">log_p_value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>colour <span class="op">=</span> <span class="va">is_sig</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ylab</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span><span class="op">(</span><span class="op">-</span><span class="va">log</span><span class="op">[</span><span class="fl">10</span><span class="op">]</span><span class="op">(</span><span class="va">p</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">3</span>, lty <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_color_manual</span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#404040"</span>, <span class="st">"#d7191c"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"B"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,
        legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span>
<span class="co">#&gt; Error in FUN(X[[i]], ...): object 'feature' not found</span></code></pre></div>
<div class="figure">
<img src="figure/2_5-1.png" alt=""><p class="caption">plot of chunk 2_5</p>
</div>
<div id="fixing-the-problem" class="section level2">
<h2 class="hasAnchor">
<a href="#fixing-the-problem" class="anchor"></a>Fixing the Problem</h2>
<p>The process above, in which we first cluster the data and then perform differential analysis, is an example of <em>double dipping</em>.</p>
<details><summary><b>What is Double Dipping?</b></summary> Double dipping arises when insufficient care is taken in data analysis that involves multiple stages of computational work. While there are at least three well-documented instances of the phenomenon (<a href="https://www.nature.com/articles/nn.2303">Kriegeskorte et al., 2009</a>), here we focus on a two-stage pipeline involving clustering and differential analysis. We dip into the data first by clustering, and then dip again by performing differential analysis on the post-clustering data. This is problematic because the clustering method inherently searches for discriminative features to determine clusters. Thus we should not dip twice, or “double dip.”
</details><p>Fortunately, we can avoid risking double dipping by performing our exchangeability test on the dataset <span class="math inline">\(\mathbf{X}\)</span> as a first step.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## How to avoid risking double dipping</span>
<span class="co"># Compute p-value for test of homogeneity</span>
<span class="va">the_p_value</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPValue.html">getPValue</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"The p-value for the exchangeability test is "</span>, <span class="va">the_p_value</span>, <span class="st">".\n"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="co">#&gt; The p-value for the exchangeability test is 0.4782.</span>
<span class="kw">if</span> <span class="op">(</span><span class="va">the_p_value</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"No evidence of heterogeneity at alpha = 0.05!\n"</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">#&gt; No evidence of heterogeneity at alpha = 0.05!</span></code></pre></div>
<p>Note that alternative approaches, which we will not explore here, include selective inference (<a href="https://arxiv.org/abs/2012.02936">Gao et al., 2021+</a>). For this alternative approach, in the differential analysis step one computes a <span class="math inline">\(p\)</span>-value <em>conditioned on the cluster membership</em> obtained from the previous clustering step.</p>
</div>
</div>
<div id="dependent-features-example-from-population-genetics" class="section level1">
<h1 class="hasAnchor">
<a href="#dependent-features-example-from-population-genetics" class="anchor"></a>Dependent Features: Example from Population Genetics</h1>
<p>The basic example, while illustrative of the problem and easily reproducible, is unlikely to occur in practice. Thus, we consider a more complicated example motivated by population genetics.</p>
<details><summary><b>Tell me More about Population Genetic Data!</b></summary> In population-genetic datasets, the features are binarized single nucleotide polymorphisms (SNPs). These features are thought to be generated by an underlying model of evolution, which includes both molecular processes (e.g., base pair mutation and genetic recombination) and demographic events (bottlenecks, migrations, etc.). Broadly speaking, a population geneticist takes a dataset <span class="math inline">\(\mathbf{X}\)</span>, whose samples <span class="math inline">\(\mathbf{x}_1,\ldots,\mathbf{x}_N\)</span> are called haplotypes or genotypes depending on the ploidy of the organism, and fits a parametric model to it to estimate biologically or demographically meaningful parameters. Such parammeters include the per-site-per-generation mutation rate <span class="math inline">\(\mu\)</span>, the per-site-per-generation recombination rate <span class="math inline">\(\rho\)</span>, and the effective population size trajectory <span class="math inline">\(N_e(t)\)</span>, just to name a few.
</details><p>Here, we simulate <span class="math inline">\(\mathbf{X}\)</span>, a dataset consisting of <span class="math inline">\(N=100\)</span> haploid individuals sampled from a panmictic population with per-site-per-generation parameters <span class="math inline">\(\mu=\rho=2\times 10^{-8}\)</span>. We generate polymorphisms by sampling <span class="math inline">\(10\)</span> chromosomes independently. For each chromosome, we assume its length is set to <span class="math inline">\(1\times 10^{5}\)</span>. The effective population size is assumed constant <span class="math inline">\(N_e=10000\)</span> throughout.</p>
<p>To simulate <span class="math inline">\(\mathbf{X}\)</span>, we use <code>msprime</code>. The simulation script is saved as <code>msprime_model.py</code>, and we call it using the <code>reticulate</code> library that interfaces R with Python.</p>
<p>Note that because of the way in which mutations are generated (they are assigned on top of a generated tree topology), the number of binary features contributed by each chromosome differs. We will treat each chromosome as an independent block later, so this remark is to remind us that our blocks will not typically be of the same size.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Generate some data</span>
<span class="co"># Source function to generate arrays with msprime</span>
<span class="fu">reticulate</span><span class="fu">::</span><span class="fu">source_python</span><span class="op">(</span><span class="st">'msprime_model.py'</span><span class="op">)</span>
<span class="co">#&gt; Error in py_run_file_impl(file, local, convert): ModuleNotFoundError: No module named 'msprime'</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Detailed traceback:</span>
<span class="co">#&gt;   File "&lt;string&gt;", line 13, in &lt;module&gt;</span>
<span class="co">#&gt;   File "/Library/Frameworks/R.framework/Versions/4.0/Resources/library/reticulate/python/rpytools/loader.py", line 44, in _import_hook</span>
<span class="co">#&gt;     level=level</span>

<span class="co"># Helper function to generate a haplotype array with names</span>
<span class="va">getExHaplotypes</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">N</span>, <span class="va">num_blocks</span><span class="op">)</span> <span class="op">{</span>
  <span class="co"># Build the matrix to be returned</span>
  <span class="va">data_set</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="va">N</span>,
                      ncol <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
  
  <span class="co"># Build the column names to be added later</span>
  <span class="va">col_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">)</span>
  <span class="co"># Build the block delimiter vector to be returned</span>
  <span class="va">block_bounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
  
  <span class="co"># Modify the matrix with simulated msprime data</span>
  <span class="kw">for</span> <span class="op">(</span><span class="va">b</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">num_blocks</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">new_subarray</span> <span class="op">&lt;-</span> <span class="fu">getHaplotypes</span><span class="op">(</span><span class="va">N</span>, <span class="fl">1e5</span><span class="op">)</span>
    <span class="kw">if</span> <span class="op">(</span><span class="va">b</span> <span class="op">&lt;</span> <span class="va">num_blocks</span><span class="op">)</span> <span class="op">{</span>
      <span class="va">block_bounds</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">block_bounds</span>, 
                         <span class="va">block_bounds</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">block_bounds</span><span class="op">)</span><span class="op">]</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">new_subarray</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>
    <span class="op">}</span>
    <span class="va">data_set</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">data_set</span>, <span class="va">new_subarray</span><span class="op">)</span>
    <span class="va">col_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">col_names</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"rs"</span>, <span class="va">b</span>, <span class="st">"_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">new_subarray</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
  <span class="op">}</span>
  
  <span class="co"># Include column names for dataset</span>
  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">data_set</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">col_names</span>
  
  <span class="co"># Store output as a list</span>
  <span class="va">to_return</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">data_set</span>,
                    bounds <span class="op">=</span> <span class="va">block_bounds</span><span class="op">)</span>
  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">to_return</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># Generate </span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">2021</span><span class="op">)</span>
<span class="va">X_data</span> <span class="op">&lt;-</span> <span class="fu">getExHaplotypes</span><span class="op">(</span>N <span class="op">=</span> <span class="fl">100</span>, num_blocks <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="co">#&gt; Error in getHaplotypes(N, 1e+05): could not find function "getHaplotypes"</span>

<span class="co"># View</span>
<span class="va">X_data</span><span class="op">$</span><span class="va">data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>
<span class="co">#&gt; Error in eval(expr, envir, enclos): object 'X_data' not found</span></code></pre></div>
<p>As in the basic example, we cluster the samples into two clusters. Here, we use Prevosti’s distance (<a href="https://link.springer.com/content/pdf/10.1007/BF00831894.pdf">Prevosti et al., 1975</a>), a popular measure of genetic distance, and perform hierarchical clustering with Ward linkage.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Perform clustering</span>
<span class="co"># Load library</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">poppr</span><span class="op">)</span>

<span class="co"># Compute Prevosti's distance matrix</span>
<span class="va">sample_dist</span> <span class="op">&lt;-</span> <span class="fu">provesti.dist</span><span class="op">(</span><span class="va">X_data</span><span class="op">$</span><span class="va">data</span><span class="op">)</span>
<span class="co">#&gt; Error in is(x, "gen"): object 'X_data' not found</span>

<span class="co"># Perform hierarchical clustering</span>
<span class="va">samples_hclust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="va">sample_dist</span>, method <span class="op">=</span> <span class="st">"ward.D"</span><span class="op">)</span>
<span class="co">#&gt; Error in hclust(sample_dist, method = "ward.D"): object 'sample_dist' not found</span>

<span class="co"># Obtain cluster labels</span>
<span class="va">k</span> <span class="op">=</span> <span class="fl">2</span>
<span class="va">samples_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html">cutree</a></span><span class="op">(</span><span class="va">samples_hclust</span>, <span class="va">k</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'as.data.frame': object 'samples_hclust' not found</span>
<span class="va">samples_clusters</span><span class="op">$</span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">1</span>, to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">X_data</span><span class="op">$</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, by <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; Error in seq.default(from = 1, to = dim(X_data$data)[1], by = 1): object 'X_data' not found</span>

<span class="co"># Attach cluster labels</span>
<span class="va">X_df</span> <span class="op">&lt;-</span> <span class="va">X_data</span><span class="op">$</span><span class="va">data</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>cluster <span class="op">=</span> <span class="va">samples_clusters</span><span class="op">$</span><span class="va">.</span><span class="op">)</span>
<span class="co">#&gt; Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'as.data.frame': object 'X_data' not found</span></code></pre></div>
<p>Similar to Section 2, we can visualize our clusters by plotting the projection of the samples of <span class="math inline">\(\mathbf{X}\)</span> onto their first two PCs. Here, because the number of features is larger than the number of samples, we use Q-PCA instead of R-PCA, and we subsequently obtain the projection scores by right-multiplying <span class="math inline">\(\mathbf{X}\)</span> with the rotation matrix.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Visualize clustering</span>
<span class="co"># Perform PCA</span>
<span class="va">X_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span><span class="va">X_df</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">X_df</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">]</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="va">X_score_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="va">X_df</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">X_df</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">%*%</span> <span class="va">X_pca</span><span class="op">$</span><span class="va">rotation</span>
<span class="va">PC1and2</span> <span class="op">&lt;-</span> <span class="va">X_score_matrix</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">PC1and2</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="va">X_df</span><span class="op">$</span><span class="va">cluster</span> 

<span class="co"># Generate plot</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">PC1and2</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span>,
             shape <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">xlab</span><span class="op">(</span><span class="st">"PC1"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ylab</span><span class="op">(</span><span class="st">"PC2"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>colour<span class="op">=</span><span class="fu">guide_legend</span><span class="op">(</span>title<span class="op">=</span><span class="st">"Cluster"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"C"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="figure/3_3-1.png" alt=""><p class="caption">plot of chunk 3_3</p>
</div>
<p>Next, we perform the Wald-Wolfowitz runs test for each feature, applying Bonferroni correction to control the family-wise error rate (FWER) at <span class="math inline">\(\alpha=0.05\)</span>.</p>
<details><summary><b>What is the Wald-Wolfowitz runs test?</b></summary> The runs test is a non-parametric test that counts the number of changes in a binary sequence (defined as either <span class="math inline">\(0\rightarrow 1\)</span> or <span class="math inline">\(1\rightarrow 0\)</span>) to decide if the sequence “looks random.” It is non-parametric because the null distribution, based on which the <span class="math inline">\(p\)</span>-value is computed, fixes the number of ones in the sequence and performs permutations of the positions of ones. For example, suppose a length <span class="math inline">\(10\)</span> binary sequence with <span class="math inline">\(4\)</span> ones is observed. If that sequence were <span class="math inline">\(1001010010\)</span>, it’s probably random. If that sequence were <span class="math inline">\(1111000000\)</span>, probably not.
</details><p>Controlling FWER with the Bonferroni method amounts to a null rejection at threshold <span class="math inline">\(0.05/P\)</span>, where <span class="math inline">\(P\)</span> is the number of polymorphic sites / features in <span class="math inline">\(\mathbf{X}\)</span>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Perform differential analysis</span>
<span class="co"># Split into two clusters </span>
<span class="va">X_df_1</span> <span class="op">&lt;-</span> <span class="va">X_df</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> 
<span class="co">#&gt; Error in filter(., cluster == 1): object 'cluster' not found</span>
<span class="va">X_df_2</span> <span class="op">&lt;-</span> <span class="va">X_df</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">cluster</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span> 
<span class="co">#&gt; Error in filter(., cluster == 2): object 'cluster' not found</span>

<span class="va">num_feat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">X_df_1</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">-</span> <span class="fl">1</span>
<span class="co">#&gt; Error in eval(expr, envir, enclos): object 'X_df_1' not found</span>

<span class="co"># Compute p-values</span>
<span class="va">p_val_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">num_feat</span><span class="op">)</span> <span class="op">{</span>
  <span class="co"># Perform differential analysis on feature j</span>
  <span class="va">da_result</span> <span class="op">&lt;-</span> <span class="fu">tseries</span><span class="fu">::</span><span class="fu">runs.test</span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">X_df_1</span><span class="op">[</span>,<span class="va">j</span><span class="op">]</span>,y <span class="op">=</span> <span class="va">X_df_2</span><span class="op">[</span>,<span class="va">j</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
  <span class="co"># Print result if significant</span>
  <span class="va">p_val_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">p_val_vector</span>, <span class="va">da_result</span><span class="op">$</span><span class="va">p.value</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">#&gt; Error in eval(expr, envir, enclos): object 'num_feat' not found</span>
<span class="va">n_sig_vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">p_val_vector</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">/</span> <span class="va">num_feat</span><span class="op">)</span>
<span class="co">#&gt; Error in eval(expr, envir, enclos): object 'num_feat' not found</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"The number of significant variants at FWER = 0.05 is "</span>, <span class="va">n_sig_vars</span>, <span class="st">"."</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="co">#&gt; Error in cat("The number of significant variants at FWER = 0.05 is ", : object 'n_sig_vars' not found</span></code></pre></div>
<p>We find many features that are statistically significant. We can visualize them using a Manhattan plot, as usual.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Visualize differential analysis</span>
<span class="co"># Mutate the p-value vector for easier plotting</span>
<span class="va">p_val_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">p_val_vector</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X_df</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="va">num_feat</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">`colnames&lt;-`</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"p_value"</span>,<span class="st">"feature"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">select</span><span class="op">(</span><span class="va">feature</span>, <span class="va">p_value</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>log_p_value <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p_value</span><span class="op">)</span>,
         is_sig <span class="op">=</span> <span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p_value</span><span class="op">)</span> <span class="op">&gt;</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">1.2e-05</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> 
<span class="co">#&gt; Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'colnames&lt;-': error in evaluating the argument 'x' in selecting a method for function 'colnames': object 'num_feat' not found</span>

<span class="co"># Generate plot</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">p_val_vector</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">feature</span>, y <span class="op">=</span> <span class="va">log_p_value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>colour <span class="op">=</span> <span class="va">is_sig</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ylab</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span><span class="op">(</span><span class="op">-</span><span class="va">log</span><span class="op">[</span><span class="fl">10</span><span class="op">]</span><span class="op">(</span><span class="va">p</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">/</span><span class="va">num_feat</span><span class="op">)</span>, lty <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_color_manual</span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#404040"</span>, <span class="st">"#d7191c"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"D"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,
        legend.position <span class="op">=</span> <span class="st">"none"</span>,
        axis.text.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
        axis.ticks.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Error in new_data_frame(list(yintercept = yintercept)): object 'num_feat' not found</span></code></pre></div>
<p>In a population-genetic setting, these “significant features” would be candidate SNPs for ancestrally informative markers, which we know to be false given that the samples were really drawn from a single, panmictic population.</p>
<div id="failing-to-consider-feature-dependencies" class="section level2">
<h2 class="hasAnchor">
<a href="#failing-to-consider-feature-dependencies" class="anchor"></a>Failing to Consider Feature Dependencies</h2>
<p>Like in the basic example, we apply our exchangeability test before performing clustering to avoid double dipping. However, for population-genetic datasets, we expect correlation between SNPs lying in the same chromosome (owing to genetic recombination). This feature-feature correlation induces correlational patterns between samples, which could falsely suggest the presence of heterogeneous communities. We can see this by running the independent features version of the test.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Compute p-value under independent features null</span>
<span class="fu"><a href="../reference/getPValue.html">getPValue</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">X_df</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">X_df</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, largeP <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.4220238</span></code></pre></div>
<p>If we were naïve, we’d reject the null hypothesis and conclude incorrectly the presence of distinct subpopulations.</p>
</div>
<div id="accounting-for-feature-dependencies" class="section level2">
<h2 class="hasAnchor">
<a href="#accounting-for-feature-dependencies" class="anchor"></a>Accounting for Feature Dependencies</h2>
<p>To apply our test correctly, we must account for feature-feature dependencies. In our case, our features are grouped into chromosome blocks, where within each chromosome the SNPs are correlated owing to genetic recombination.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Compute p-value under independent blocks null</span>
<span class="co"># Note: this takes about 2-3 minutes to run</span>
<span class="fu"><a href="../reference/getPValue.html">getPValue</a></span><span class="op">(</span><span class="va">X_data</span><span class="op">$</span><span class="va">data</span>, 
          block_boundaries <span class="op">=</span> <span class="va">X_data</span><span class="op">$</span><span class="va">bounds</span><span class="op">)</span>
<span class="co">#&gt; Error in eval(assertion, env): object 'X_data' not found</span></code></pre></div>
<p>Now, we obtain a <span class="math inline">\(p\)</span>-value that is well above a typical cut-off threshold. Hence, we conclude correctly that at <span class="math inline">\(\alpha=0.05\)</span> there is insufficient evidence that <span class="math inline">\(\mathbf{X}\)</span> is heterogeneous.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Alan Aw.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
