<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of scATAC-seq Data ‚Ä¢ flintyR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of scATAC-seq Data">
<meta property="og:description" content="flintyR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">flintyR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ex_vs_hom.html">Exchangeability and Homogeneity</a>
    </li>
    <li>
      <a href="../articles/extras.html">Extras</a>
    </li>
    <li>
      <a href="../articles/intro.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/single-cell-atac-seq.html">Analysis of scATAC-seq Data</a>
    </li>
    <li>
      <a href="../articles/wvs.html">Analysis of World Values Survey Data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/alanaw1/flintyR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="single-cell-atac-seq_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysis of scATAC-seq Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/alanaw1/flintyR/blob/master/vignettes/single-cell-atac-seq.Rmd"><code>vignettes/single-cell-atac-seq.Rmd</code></a></small>
      <div class="hidden name"><code>single-cell-atac-seq.Rmd</code></div>

    </div>

    
    
<p><em>We are grateful to <a href="https://timoast.github.io/">Tim Stuart</a>, Postdoctoral Researcher at the New York Genome Center, for feedback on accuracy, enhancing accessibility, and improving reproducibility of this vignette.</em></p>
<p>In cellular biology, the study of single cells allows biologists to discover mechanisms not detectable by studying only bulk populations of cells. In recent years, advances in sequencing technologies have enabled the quantification of many single cell features (e.g., transcriptome, proteome, metabolome, 3D contact maps), and such datasets have in turn provided new statistical challenges and spurred the development of new probabilistic models (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-1926-6">L√§hnemann et al., 2020</a>).</p>
<p>Key to discovering mechanisms from single cell data are analyses of heterogeneity between single cell populations. Heterogeneous patterns across multiple biological features reveal biological differences between groups of cells that differ by physical location, temporal factors, or some other biological function, and such differences can subsequently help generate hypotheses relating features to function. Moreover, depending on the question the resolution of the population might differ, for instance at the tissue level, at the cell type level, or even at physical location level among the same cell types.</p>
<p style="text-align:center;">
<img src="ex_data/scATAC-seq/bmc_fig1.png" alt="Many resolutions of single cell statistical inference." width="500"></p>
<p align="center">
<em>Many levels of resolution at which to investigate single cells. Figure reproduced from <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-1926-6">L√§hnemann et al.¬†(2020)</a> and <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1663-x">Wolf et al.¬†(2019)</a> under the Creative Commons Attribution (CC-BY) license.</em>
</p>
<p>Central to single cell analyses is the clustering of a large population of sampled cells from an experiment. Such clustering often partitions the large population into single cell communities that correspond to some known biological function (e.g., cell type or sampling batch). Once these cell communities are obtained, further analyses can be conducted <em>on</em> or <em>between</em> identified communities to generate biological hypotheses.</p>
<p>üìñ In this vignette, we will:</p>
<ul>
<li>show how <strong>flinty</strong> can be applied to detect statistical heterogeneity in single cell clusters detected from scATAC-seq data,</li>
<li>demonstrate how <strong>flinty</strong> can explain drivers of the detected heterogeneity.</li>
</ul>
<div id="single-cell-atac-seq-data" class="section level1">
<h1 class="hasAnchor">
<a href="#single-cell-atac-seq-data" class="anchor"></a>Single Cell ATAC-seq Data</h1>
<p>We obtain publicly available human peripheral blood mononuclear cell (PBMC) scATAC-seq data from <a href="https://www.10xgenomics.com/">10X Genomics</a>. Data processing requires <strong>Seurat</strong> and its extension, <strong>Signac (v1.3.0)</strong>; both can be installed with Bioconductor.</p>
<p>The code below loads and describes the PBMC data <em>before</em> running data processing steps.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Install older version of Signac</span>
<span class="co"># devtools::install_version("Signac", version = "0.2.5")</span>

<span class="co">## Load packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Signac</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Seurat</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v75</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">patchwork</span><span class="op">)</span>

<span class="co">## Load the original data</span>
<span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">Read10X_h5</span><span class="op">(</span>filename <span class="op">=</span> <span class="st">"ex_data/scATAC-seq/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5"</span><span class="op">)</span>
<span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span>
  file <span class="op">=</span> <span class="st">"ex_data/scATAC-seq/atac_v1_pbmc_10k_singlecell.csv"</span>,
  header <span class="op">=</span> <span class="cn">TRUE</span>,
  row.names <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span>

<span class="va">chrom_assay</span> <span class="op">&lt;-</span> <span class="fu">CreateChromatinAssay</span><span class="op">(</span>
  counts <span class="op">=</span> <span class="va">counts</span>,
  sep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">":"</span>, <span class="st">"-"</span><span class="op">)</span>,
  genome <span class="op">=</span> <span class="st">'hg19'</span>,
  fragments <span class="op">=</span> <span class="st">'ex_data/scATAC-seq/atac_v1_pbmc_10k_fragments.tsv.gz'</span>,
  min.cells <span class="op">=</span> <span class="fl">10</span>,
  min.features <span class="op">=</span> <span class="fl">200</span>
<span class="op">)</span>

<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">CreateSeuratObject</span><span class="op">(</span>
  counts <span class="op">=</span> <span class="va">chrom_assay</span>,
  assay <span class="op">=</span> <span class="st">"peaks"</span>,
  meta.data <span class="op">=</span> <span class="va">metadata</span>
<span class="op">)</span>

<span class="co">## Show the PBMC object before data processing</span>
<span class="va">pbmc</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 87561 features across 8728 samples within 1 assay </span>
<span class="co">#&gt; Active assay: peaks (87561 features, 0 variable features)</span></code></pre></div>
<div id="data-processing" class="section level2">
<h2 class="hasAnchor">
<a href="#data-processing" class="anchor"></a>Data Processing</h2>
<p>We perform all steps of data processing up to <a href="https://satijalab.org/signac/articles/pbmc_vignette.html#non-linear-dimension-reduction-and-clustering-1">Non-linear dimension reduction and clustering</a> in this <a href="https://satijalab.org/signac/articles/pbmc_vignette.html">tutorial</a> provided by the lab of Rahul Satija.</p>
<p>To summarize the essential steps, we</p>
<ol style="list-style-type: decimal">
<li>Filter individual cells that fail to satisfy criteria associated with five quality control metrics (nucleosome banding pattern, TSS enrichment score, total number of fragments in peaks, fraction of fragments in peaks and ratio reads in blacklist sites provided by ENCODE);</li>
<li>Perform term frequency-inverse document frequency (TF-IDF) normalization on the peak reads matrix;</li>
<li>Perform singular value decomposition (SVD) on the TF-IDF normalized matrix to reduce dimensionality;</li>
<li>Cluster the cells using the <a href="https://link.springer.com/article/10.1140/epjb/e2013-40829-0">smart local moving (SLM)</a> algorithm.</li>
</ol>
<p>We obtain <span class="math inline">\(7060\)</span> cells after Step 1, and Step 4 yielded <span class="math inline">\(15\)</span> clusters altogether, whose sizes are reported below. Steps 2 and 3 together account for technical noise (e.g., differences in sequencing depth). This step is important for us because we wish to interrogate homogeneity of clusters and to explain any <em>remaining</em> sources of heterogeneity in the data <em>not</em> owing to technical noise, if there are any.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load summary of clustering, show cluster sizes</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"ex_data/scATAC-seq/scATAC-seq_annotation_cluster.RData"</span><span class="op">)</span>
<span class="fu">kableExtra</span><span class="fu">::</span><span class="fu">kable</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">combined_metadata</span><span class="op">$</span><span class="va">umap_cluster_labels</span><span class="op">)</span>,
                  col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Cluster"</span>, <span class="st">"Count"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">kableExtra</span><span class="fu">::</span><span class="fu">kable_styling</span><span class="op">(</span>position <span class="op">=</span> <span class="st">"center"</span><span class="op">)</span></code></pre></div>
<table class="table table">
<thead><tr>
<th style="text-align:left;">
Cluster
</th>
<th style="text-align:right;">
Count
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
0
</td>
<td style="text-align:right;">
2530
</td>
</tr>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:right;">
722
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:right;">
662
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:right;">
587
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:right;">
578
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:right;">
398
</td>
</tr>
<tr>
<td style="text-align:left;">
6
</td>
<td style="text-align:right;">
357
</td>
</tr>
<tr>
<td style="text-align:left;">
7
</td>
<td style="text-align:right;">
282
</td>
</tr>
<tr>
<td style="text-align:left;">
8
</td>
<td style="text-align:right;">
244
</td>
</tr>
<tr>
<td style="text-align:left;">
9
</td>
<td style="text-align:right;">
234
</td>
</tr>
<tr>
<td style="text-align:left;">
10
</td>
<td style="text-align:right;">
202
</td>
</tr>
<tr>
<td style="text-align:left;">
11
</td>
<td style="text-align:right;">
128
</td>
</tr>
<tr>
<td style="text-align:left;">
12
</td>
<td style="text-align:right;">
60
</td>
</tr>
<tr>
<td style="text-align:left;">
13
</td>
<td style="text-align:right;">
50
</td>
</tr>
<tr>
<td style="text-align:left;">
14
</td>
<td style="text-align:right;">
26
</td>
</tr>
</tbody>
</table>
<p>These clusters can be visualized in 2D space using UMAP. The code to produce this UMAP plot takes some time to run, so we avoid running it here. Interested readers can find the code in the Satija lab vignettes linked above.</p>
<p style="text-align:center;">
<img src="ex_data/scATAC-seq/pbmc_umap_plot.jpg" alt="UMAP showing separation of clusters." width="400"></p>
<p>These clusters correspond to myeloid and lymphoid cell populations. One can identify cell types by identifying cluster-defining biological markers using RNA-seq measurements of the same cells; see the Satija lab vignettes <a href="https://satijalab.org/signac/articles/pbmc_vignette.html#integrating-with-scrna-seq-data-1">here</a> and <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html">there</a> for details on accomplishing this ‚Äúfeature anchoring‚Äù step.</p>
<p style="text-align:center;">
<img src="ex_data/scATAC-seq/pbmc_pie_chart.jpg" alt="Pie charts summarizing cell type composition per cluster" width="800"></p>
<p>Before proceeding with our analysis, we note that the <span class="math inline">\(87561\)</span> features per individual cell correspond to physical regions along the human genome, and each feature records the number of reads, more precisely Tn5cut sites, mapped by that individual cell to the physical region. These read-based records measure chromatin activity and thus provide information about gene regulation.</p>
</div>
</div>
<div id="analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis" class="anchor"></a>Analysis</h1>
<p>Data processing has provided us with a dataset <span class="math inline">\(\mathbf{X}\in\mathbb{R}^{N\times P}\)</span> consisting of <span class="math inline">\(N=7060\)</span> single cells with <span class="math inline">\(P=87561\)</span> TF-IDF normalized peak reads/counts features recorded per cell. Additionally, <span class="math inline">\(\mathbf{X}\)</span> has been partitioned into <span class="math inline">\(15\)</span> clusters that correspond to certain distinguishable blood cell populations.</p>
<p>Are the clusters homogeneous? How can one verify so? A na√Øve approach is to assume the normalized peak reads are statistically independent from one another and check for heterogeneity. We do this below for Cluster 11, running the independent features version of our exchangeability test and restricting to peak reads lying in Chromosome 1 to save time. For convenience, we have also saved the clusters generated above, so that we can load them directly and run our test.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Show Cluster 11 (first six cells and first six features) </span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"cluster_11_pbmc_data.RData"</span><span class="op">)</span>
<span class="co"># What is the dimension?</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">extracted_matrix</span><span class="op">)</span>
<span class="co">#&gt; [1]   128 87561</span>
<span class="co"># Preview first six cells and six normalized features</span>
<span class="va">extracted_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>
<span class="co">#&gt; 6 x 6 sparse Matrix of class "dgCMatrix"</span>
<span class="co">#&gt;                    chr1-565107-565550 chr1-569174-569639 chr1-713460-714823</span>
<span class="co">#&gt; AAACTGCTCCTATCCG-1                  .                  .           1.746733</span>
<span class="co">#&gt; AAATGCCAGTGAAACT-1                  .                  .           .       </span>
<span class="co">#&gt; AACCAACAGTGTCGGA-1                  .                  .           .       </span>
<span class="co">#&gt; AACCTGATCGCTAGTA-1                  .                  .           .       </span>
<span class="co">#&gt; AACGGGATCTTGTCGC-1                  .                  .           .       </span>
<span class="co">#&gt; AAGATAGGTGCTTACA-1                  .                  .           .       </span>
<span class="co">#&gt;                    chr1-752422-753038 chr1-762106-763359 chr1-779589-780271</span>
<span class="co">#&gt; AAACTGCTCCTATCCG-1                  .                  .                  .</span>
<span class="co">#&gt; AAATGCCAGTGAAACT-1                  .                  .                  .</span>
<span class="co">#&gt; AACCAACAGTGTCGGA-1                  .                  .                  .</span>
<span class="co">#&gt; AACCTGATCGCTAGTA-1                  .                  .                  .</span>
<span class="co">#&gt; AACGGGATCTTGTCGC-1                  .                  .                  .</span>
<span class="co">#&gt; AAGATAGGTGCTTACA-1                  .                  .                  .</span>
<span class="co"># Get peak locations</span>
<span class="va">peak_locs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">extracted_matrix</span><span class="op">)</span>
<span class="co"># Extract chromosome number</span>
<span class="va">chr_block_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">peak_locs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/substr.html">substring</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span><span class="op">(</span><span class="va">x</span>, split <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="fl">4</span><span class="op">)</span><span class="op">}</span><span class="op">)</span>
<span class="co"># Extract position of all chr1 chromosomes</span>
<span class="va">last_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="st">"2"</span>,<span class="va">chr_block_labels</span><span class="op">)</span>
<span class="va">last_id</span> <span class="op">&lt;-</span> <span class="va">last_id</span> <span class="op">-</span> <span class="fl">1</span>
<span class="co"># Restrict to peak reads in Chr1</span>
<span class="va">X</span> <span class="op">&lt;-</span> <span class="va">extracted_matrix</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="va">last_id</span><span class="op">]</span>
<span class="co"># Check dimension</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span>
<span class="co">#&gt; [1]  128 8555</span>

<span class="co">## Run test of exchangeability</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://alanaw1.github.io/flintyR/">flintyR</a></span><span class="op">)</span>
<span class="fu"><a href="../reference/getPValue.html">getPValue</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span>, largeP <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># use large P asymptotics since P is large</span>
<span class="co">#&gt; Warning in max(block_labels): no non-missing arguments to max; returning -Inf</span>
<span class="co">#&gt; Warning in seq_len(length.out - 2L) * by: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
<span class="co">#&gt;   Use c() or as.vector() instead.</span>
<span class="co">#&gt; Warning in val - grid: Recycling array of length 1 in array-vector arithmetic is deprecated.</span>
<span class="co">#&gt;   Use c() or as.vector() instead.</span>
<span class="co">#&gt; [1] 4.260687e-73</span></code></pre></div>
<p>The <span class="math inline">\(p\)</span>-value above provides evidence that Cluster 11 is non-exchangeable. However, it could be a Type I error arising from wrongly assuming that the features are statistically independent. (In other words, it could also be interpreted as evidence for dependencies between peaks. It could also be interpreted as evidence for <em>both</em> cluster heterogeneity and peak dependency.)</p>
<p>To account for statistical dependencies between peaks, we recall some possible biological factors inducing correlations between peaks.</p>
<ul>
<li>Transposase-accessible chromatin peaks map chromatin accessibility, which is typically regulated by transcription factor (TF) binding; see <a href="https://epigeneticsandchromatin.biomedcentral.com/articles/10.1186/1756-8935-7-33">Tsompana and Buck (2014)</a>. Gene regulatory networks that govern the rules of binding induce correlations between peaks, via specific mechanisms through which the associated TFs interact with chromatin.</li>
<li>Recent studies have explored how chromosome conformation, or chromatin folding, may implicate gene regulation. Markov models using chromatin-chromatin interaction data (measured by chromosome conformation capture methods like Hi-C) can predict the distributions of activation histone modifications and transcription factors (<a href="https://doi.org/10.1093/nar/gkx086">Wang et al., 2017</a>), and the precise dynamic relationship can be complicated (<a href="https://www.nature.com/articles/s41467-018-04130-x">Cortini and Filion, 2018</a>) and also potentially highly contingent on background cellular processes (<a href="https://www.nature.com/articles/s41467-019-09483-5">Bertero and Fields et al., 2019</a>).</li>
</ul>
<p>In our paper and in this vignette, we attempt to investigate the second point by leveraging Hi-C data to group peaks in the scATAC-seq data. Owing to the large number of peak regions (<span class="math inline">\(P=87561\)</span>) in our scATAC-seq data, for our analysis we will restrict our features to only those peak regions that lie in Chromosome 1. This restriction reduces the number of features to <span class="math inline">\(P=8555\)</span>, as we saw when we ran our test on Cluster 11 earlier.</p>
<div id="grouping-peaks-with-hi-c-data" class="section level2">
<h2 class="hasAnchor">
<a href="#grouping-peaks-with-hi-c-data" class="anchor"></a>Grouping Peaks with Hi-C Data</h2>
<p>Before proceeding to integrate Hi-C data with our scATAC-seq data, we provide some background on Hi-C. As described in the second bullet point above, there is evidence suggesting that the 3D structure of the genome plays an important role in determining differences in chromatin accessibility. Chromatin 3D organization is measured by probing physical interaction between non-adjacent genomic loci and such probing in turn is accomplished by chromatin conformation capture (3C) technologies (<a href="https://science.sciencemag.org/content/295/5558/1306">Dekker et al., 2002</a>). With Hi-C (<a href="https://science.sciencemag.org/content/326/5950/289">Lieberman-Aiden et al., 2009</a>) being one of the more advanced capturing technologies, we hence use Hi-C data to supervise the grouping of our peaks. The heat map below depicts what a Hi-C interaction map looks like.</p>
<p style="text-align:center;">
<img src="ex_data/scATAC-seq/GM06990.jpg" alt="Hi-C Contact Map Reads." width="400"></p>
<p align="center">
<em>Scatterplot of a Hi-C interaction map of Chromosome 14 of a genome sequenced from <a href="https://www.coriell.org/0/Sections/Search/Sample_Detail.aspx?Ref=GM06990">lymphoblastoid cells (GM06990)</a>. Localized darker patches reveal local regions of increased contact. Plot produced using the Bioconductor package <a href="https://bioconductor.org/packages/release/data/experiment/html/LiebermanAidenHiC2009.html">LiebermanAidenHiC2009</a> with help from <a href="https://bioconductor.riken.jp/packages/3.8/data/experiment/vignettes/LiebermanAidenHiC2009/inst/doc/LiebermanAidenHiC2009.pdf">this vignette</a>.</em>
</p>
<p>Concretely, we use Hi-C data to partition our features into disjoint sets of dependent features (so-called ‚Äúdependent blocks‚Äù), where we group features into the same sets based on whether they belong in the same topologically associating domain (TAD) as identified by Hi-C TAD callers.</p>
</div>
<div id="methods-to-call-tads-on-hi-c-data" class="section level2">
<h2 class="hasAnchor">
<a href="#methods-to-call-tads-on-hi-c-data" class="anchor"></a>Methods to Call TADs on Hi-C Data</h2>
<p>There are more than 20 TAD calling approaches; see <a href="https://link.springer.com/article/10.1007/s12551-018-0489-1">Pal et al.¬†(2019)</a> for a summary of them. We consider two approaches, <strong>DomainCaller</strong> ‚Äî a method proposed by <a href="https://www.nature.com/articles/nature11082">Dixon et al.¬†(2012)</a> that relies on the directionality index ‚Äî and <strong>TopDom</strong> ‚Äî a newer approach that scores patterns of contact frequencies and tests these scores for statistical significance under a non-parametric null distribution <a href="https://academic.oup.com/nar/article/44/7/e70/2467818">Shin et al.¬†(2016)</a>. Note both <strong>DomainCaller</strong> and <strong>TopDom</strong> are found reliable in a comprehensive meta-analysis of 22 TAD callers by <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1596-9">Zufferey et al.¬†(2018)</a>. (TADs were first described by Dixon and colleagues, so it should come as no surprise that their <strong>DomainCaller</strong> method identifies TADs well.)</p>
<p>We provide a guide to generating TAD boundaries in a separate vignette.</p>
<details><summary><b>A Important Caveat</b></summary> Because the single cell data provided by 10X Genomics is not accompanied by Hi-C contact maps of the same cells, we instead rely on an external Hi-C dataset consisting of human fetal lung fibroblast (IfMR90) cells (<a href="https://www.nature.com/articles/nature11082">Dixon et al., 2012</a>). This introduces one caveat to our analysis: we assume TADs identified from IMR90 cells are approximately identical to TADs in PBMC. Note that there is some evidence suggesting that TADs are reasonably preserved across cell types; see <a href="https://academic.oup.com/bioinformatics/article/34/13/i475/5045717">Sauerwald and Kingsford (2018)</a> for instance. Careful quantification of the invariance of TADs toward cell type specification is a topic of ongoing research, and we direct the interested reader to recent work, e.g., <a href="https://www.nature.com/articles/s41467-018-03017-1">Gong et al.¬†(2018)</a>, <a href="https://bmcbiol.biomedcentral.com/articles/10.1186/s12915-018-0556-x">Krefting et al.¬†(2018)</a>, and <a href="https://www.sciencedirect.com/science/article/abs/pii/S000292972100001X">McArthur and Capra (2021)</a>.
</details>
</div>
</div>
<div id="results" class="section level1">
<h1 class="hasAnchor">
<a href="#results" class="anchor"></a>Results</h1>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load packages and set directories</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://Matrix.R-forge.R-project.org/">Matrix</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">doParallel</span><span class="op">)</span>
<span class="co"># Register parallel computation</span>
<span class="fu"><a href="https://rdrr.io/pkg/doParallel/man/registerDoParallel.html">registerDoParallel</a></span><span class="op">(</span><span class="op">)</span>
<span class="co"># Relative directory</span>
<span class="va">rel_dir</span> <span class="op">&lt;-</span> <span class="st">"ex_data/scATAC-seq"</span></code></pre></div>
<div id="are-clusters-homogeneous" class="section level2">
<h2 class="hasAnchor">
<a href="#are-clusters-homogeneous" class="anchor"></a>Are Clusters Homogeneous?</h2>
<p>Our first goal is to check whether each cluster of cells is homogeneous, even after accounting for potential dependencies between peaks arising from chromosome folding. To accomplish this, we need to group peaks based on whether they lie in the same TAD. We demonstrate how we perform this step, using TAD membership within Chromosome 1 identified by <strong>DomainCaller</strong>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Grouping peaks</span>
<span class="co"># Load libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">HiTC</span><span class="op">)</span> <span class="co"># install using Bioconductor if not installed</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">HiCDataHumanIMR90</span><span class="op">)</span> <span class="co"># install using Bioconductor if not installed</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">Dixon2012_IMR90</span><span class="op">)</span> 

<span class="co"># View TADs identified by DomainCaller</span>
<span class="va">tads_imr90</span>
<span class="co">#&gt; GRanges object with 2338 ranges and 0 metadata columns:</span>
<span class="co">#&gt;            seqnames              ranges strand</span>
<span class="co">#&gt;               &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="co">#&gt;      TAD-1     chr1      770138-1290137      *</span>
<span class="co">#&gt;      TAD-2     chr1     1290138-1850140      *</span>
<span class="co">#&gt;      TAD-3     chr1     1850141-2330140      *</span>
<span class="co">#&gt;      TAD-4     chr1     2330141-3610140      *</span>
<span class="co">#&gt;      TAD-5     chr1     3770141-6077413      *</span>
<span class="co">#&gt;        ...      ...                 ...    ...</span>
<span class="co">#&gt;   TAD-2334     chrX 146992309-148552096      *</span>
<span class="co">#&gt;   TAD-2335     chrX 148592096-149929342      *</span>
<span class="co">#&gt;   TAD-2336     chrX 149929343-151969344      *</span>
<span class="co">#&gt;   TAD-2337     chrX 152089345-152746806      *</span>
<span class="co">#&gt;   TAD-2338     chrX 152786807-154946806      *</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 23 sequences from an unspecified genome; no seqlengths</span>

<span class="co"># Helper function for classifying each ATAC peak into one of many TADs</span>
<span class="va">getDomainCallerLabels</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">tads_imr90</span>, <span class="va">peak_locs_2</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">chr1_tad_locs</span> <span class="op">&lt;-</span> <span class="va">tads_imr90</span><span class="op">@</span><span class="va">ranges</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">tads_imr90</span><span class="op">@</span><span class="va">seqnames</span> <span class="op">==</span> <span class="st">"chr1"</span><span class="op">)</span>,<span class="op">]</span>
  
  <span class="va">start_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">peak_locs_2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span><span class="op">(</span><span class="va">x</span>, split <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">}</span><span class="op">)</span>
  <span class="va">end_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">peak_locs_2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span><span class="op">(</span><span class="va">x</span>, split <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">}</span><span class="op">)</span>
  <span class="va">tmp_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>START <span class="op">=</span> <span class="va">start_pos</span>,
                       END <span class="op">=</span> <span class="va">end_pos</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">tmp_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
  <span class="va">tmp_df</span><span class="op">$</span><span class="va">START</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">tmp_df</span><span class="op">$</span><span class="va">START</span><span class="op">)</span>; <span class="va">tmp_df</span><span class="op">$</span><span class="va">END</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">tmp_df</span><span class="op">$</span><span class="va">END</span><span class="op">)</span>
  <span class="va">getThisLabel</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">start_</span>, <span class="va">end_</span><span class="op">)</span> <span class="op">{</span>
    <span class="co"># Find largest start in chr1_tad_locs that start_ beats</span>
    <span class="va">max_start_ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">chr1_tad_locs</span><span class="op">@</span><span class="va">start</span> <span class="op">&lt;</span> <span class="va">start_</span><span class="op">)</span><span class="op">)</span>
    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/is.finite.html">is.infinite</a></span><span class="op">(</span><span class="va">max_start_ind</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
      <span class="co"># Feature has no available block label, set to -99</span>
      <span class="fu"><a href="https://rdrr.io/r/base/message.html">message</a></span><span class="op">(</span><span class="st">"Peak Region is not contained in any TAD..."</span><span class="op">)</span>
      <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="op">-</span><span class="fl">99</span><span class="op">)</span>
    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="op">(</span><span class="va">chr1_tad_locs</span><span class="op">@</span><span class="va">start</span><span class="op">[</span><span class="va">max_start_ind</span><span class="op">]</span> <span class="op">+</span> <span class="va">chr1_tad_locs</span><span class="op">@</span><span class="va">width</span><span class="op">[</span><span class="va">max_start_ind</span><span class="op">]</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="va">end_</span><span class="op">)</span> <span class="op">{</span>
      <span class="co"># Feature is sandwiched between two TADs</span>
      <span class="fu"><a href="https://rdrr.io/r/base/message.html">message</a></span><span class="op">(</span><span class="st">"Peak Region is sandwiched between two TADs..."</span><span class="op">)</span>
      <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="op">-</span><span class="fl">99</span><span class="op">)</span>
    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span>
      <span class="fu">assertthat</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/assertthat/man/assert_that.html">assert_that</a></span><span class="op">(</span><span class="op">(</span><span class="va">chr1_tad_locs</span><span class="op">@</span><span class="va">start</span><span class="op">[</span><span class="va">max_start_ind</span><span class="op">]</span> <span class="op">+</span> <span class="va">chr1_tad_locs</span><span class="op">@</span><span class="va">width</span><span class="op">[</span><span class="va">max_start_ind</span><span class="op">]</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&gt;</span> <span class="va">end_</span>,
                              msg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"end position "</span>, <span class="va">end_</span>,<span class="st">" exceeds right end of TAD"</span>, <span class="va">chr1_tad_locs</span><span class="op">@</span><span class="va">start</span><span class="op">[</span><span class="va">max_start_ind</span>  <span class="op">+</span>  <span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
      <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">max_start_ind</span><span class="op">)</span>
    <span class="op">}</span>
  <span class="op">}</span>
  
  <span class="co"># Allocate labels</span>
  <span class="va">tmp_df</span><span class="op">$</span><span class="va">LABEL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html">mapply</a></span><span class="op">(</span><span class="va">getThisLabel</span>,<span class="va">tmp_df</span><span class="op">$</span><span class="va">START</span>, <span class="va">tmp_df</span><span class="op">$</span><span class="va">END</span><span class="op">)</span>
  
  <span class="co"># Return</span>
  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">tmp_df</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># Get domainCaller labels</span>
<span class="va">peak_locs_2</span> <span class="op">&lt;-</span> <span class="va">peak_locs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">peak_locs</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/substr.html">substring</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span><span class="op">(</span><span class="va">x</span>, split <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="fl">4</span><span class="op">)</span><span class="op">}</span><span class="op">)</span> <span class="op">==</span> <span class="st">"1"</span><span class="op">)</span><span class="op">]</span>
<span class="va">each_peaks_tad_membership</span> <span class="op">&lt;-</span> <span class="fu">getDomainCallerLabels</span><span class="op">(</span><span class="va">tads_imr90</span>, <span class="va">peak_locs_2</span><span class="op">)</span>

<span class="co"># View each peak's TAD membership</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">each_peaks_tad_membership</span><span class="op">)</span>
<span class="co">#&gt;    START    END LABEL</span>
<span class="co">#&gt; 1 565107 565550   -99</span>
<span class="co">#&gt; 2 569174 569639   -99</span>
<span class="co">#&gt; 3 713460 714823   -99</span>
<span class="co">#&gt; 4 752422 753038   -99</span>
<span class="co">#&gt; 5 762106 763359   -99</span>
<span class="co">#&gt; 6 779589 780271     1</span>

<span class="co"># How many TADs are there? </span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">each_peaks_tad_membership</span><span class="op">$</span><span class="va">LABEL</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] 233</span>

<span class="co"># Label the peaks of Cluster 11 data</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">each_peaks_tad_membership</span><span class="op">$</span><span class="va">LABEL</span></code></pre></div>
<p>Now that we have our peaks grouped by TAD membership, we can visualize correlations between groups of peaks. Visualization provides us with intuition as to whether peaks within the same group are more correlated with one another than are peaks between different groups. Below, we plot the average correlations between peaks from each pair of groups. We do not compute self-correlations when averaging correlations between peaks from the same group, and drop groups consisting of only a single peak.</p>
<p style="text-align:center;">
<img src="ex_data/scATAC-seq/ave_cor_heatmap.jpg" alt="Average correlations between groups of peaks." width="450"></p>
<p align="center">
<em>Heatmap of average correlations between groups of peaks. When computing average correlations between peaks within the same group (the diagonal elements of the heatmap), self-correlations are removed. The heatmap is generated by the code below.</em>
</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Code to generate average correlations between groups of peaks (NOT RUN HERE)</span>
<span class="co"># Obtain Chr1 peaks for entire sample of 7060 cells</span>
<span class="co">#tfidf_norm_matrix &lt;- t(pbmc@assays$peaks@data)</span>
<span class="co">#tfidf_norm_matrix_chr1 &lt;- tfidf_norm_matrix[,1:8555]</span>
<span class="co">#colnames(tfidf_norm_matrix_chr1) &lt;- each_peaks_tad_membership$LABEL</span>
<span class="co"># Delete the non-TAD classified features</span>
<span class="co">#tfidf_norm_matrix_chr1_ &lt;- tfidf_norm_matrix_chr1[,which(test_df_output$LABEL != -99)]</span>
<span class="co"># Rename block labels to be compatible </span>
<span class="co">#colnames(tfidf_norm_matrix_chr1_) &lt;- as.numeric(as.factor(colnames(tfidf_norm_matrix_chr1_)))</span>
<span class="co"># Order the peaks by groups</span>
<span class="co">#tfidf_norm_matrix_chr1_sub_mat &lt;- tfidf_norm_matrix_chr1_ %&gt;% as.matrix()</span>
<span class="co">#tfidf_norm_matrix_chr1_sub_mat &lt;- tfidf_norm_matrix_chr1_sub_mat[,order(colnames(tfidf_norm_matrix_chr1_sub_mat))]</span>
<span class="co">#all_tads &lt;- unique(colnames(tfidf_norm_matrix_chr1_sub_mat))</span>
<span class="co"># Construct average correlation matrix to plot heatmap</span>
<span class="co">#ave_cor_matrix &lt;- matrix(nrow = length(all_tads), </span>
<span class="co">#                         ncol = length(all_tads))</span>
<span class="co">#for (i in all_tads) {</span>
<span class="co">#  for (j in all_tads) {</span>
<span class="co">#    # Extract submatrix belonging to TAD i</span>
<span class="co">#    TAD_i &lt;- tfidf_norm_matrix_chr1_sub_mat[,which(colnames(tfidf_norm_matrix_chr1_sub_mat) == i),drop=FALSE]</span>
<span class="co">#    if (i == j) {</span>
<span class="co">#      # Compute the non-matching features only</span>
<span class="co">#      if (ncol(TAD_i) == 1) {</span>
<span class="co">#        x_ij &lt;- 0</span>
<span class="co">#      } else {</span>
<span class="co">#        cor_ii &lt;- cor(TAD_i)</span>
<span class="co">#        cor_ii &lt;- cor_ii - diag(1, nrow(cor_ii), ncol(cor_ii))</span>
<span class="co">#        x_ij &lt;- mean(cor_ii) * nrow(cor_ii)^2 / (nrow(cor_ii)^2 - nrow(cor_ii))</span>
<span class="co">#      }</span>
<span class="co">#    } else {</span>
<span class="co">#      # Extract submatrix belonging to TAD j</span>
<span class="co">#      TAD_j &lt;- tfidf_norm_matrix_chr1_sub_mat[,which(colnames(tfidf_norm_matrix_chr1_sub_mat) == j),drop=FALSE]</span>
<span class="co">#      # Compute average correlation between all features in TAD i and all features in TAD j</span>
<span class="co">#      x_ij &lt;- mean(cor(TAD_i, TAD_j))</span>
<span class="co">#    }</span>
<span class="co">#    # Fill in entry (i,j) of ave_cor_matrix</span>
<span class="co">#    ave_cor_matrix[as.numeric(i),as.numeric(j)] &lt;- x_ij</span>
<span class="co">#  }</span>
<span class="co">#}</span>
<span class="co"># Save average correlation matrix</span>
<span class="co">#write.csv(ave_cor_matrix, file = "ave_cor_matrix.csv")</span>
<span class="co"># Load average correlation matrix</span>
<span class="va">ave_cor_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">rel_dir</span>, <span class="st">"/ave_cor_matrix.csv"</span><span class="op">)</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="co"># Plot heat map</span>
<span class="co">#col_scheme &lt;- colorRampPalette(c("blue", "white", "red"))(30)</span>
<span class="co">#gplots::heatmap.2(as.matrix(ave_cor_matrix), col=col_scheme, dendrogram='none', Rowv=FALSE, Colv=FALSE, trace='none')</span>
<span class="co"># Compute mean diagonal and off-diagonal correlation values</span>
<span class="va">ave_cor_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">ave_cor_matrix</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">ave_cor_matrix</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.009481952</span>
<span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">ave_cor_matrix</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">ave_cor_matrix</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">ave_cor_matrix</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span> <span class="op">/</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">ave_cor_matrix</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">ave_cor_matrix</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] -0.00712446</span></code></pre></div>
<p>The heat map and calculation above suggest that peak-peak correlations within a group are slightly higher than peak-peak correlations between distinct groups. However, average correlations between pairs of groups are all very weak, never exceeding <span class="math inline">\(0.01\)</span> in magnitude.</p>
<p>Take note that TAD grouping is contiguous in nature: peaks physically close to one another are typically grouped together as a TAD, with occasional boundaries breaking the contiguity. This is similar to the grouping of single nucleotide polymorphisms in genetics into contiguous LD blocks, where within-block correlations are empirically higher than between-block correlations.</p>
<p>We proceed with running our exchangeability test. Since the number of blocks <span class="math inline">\(B=233 - 1 = 232\)</span> (one block corresponds to unclassified peaks) is large, we use the asymptotic version of our test.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Grouping peaks</span>
<span class="co"># Delete the non-TAD classified features</span>
<span class="va">X_</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">each_peaks_tad_membership</span><span class="op">$</span><span class="va">LABEL</span> <span class="op">!=</span> <span class="op">-</span><span class="fl">99</span><span class="op">)</span><span class="op">]</span>
<span class="co"># Remove TADs containing only a single peak</span>
<span class="va">X__</span> <span class="op">&lt;-</span> <span class="va">X_</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X_</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span><span class="op">]</span>
<span class="co"># Rename block labels to be compatible </span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X__</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X__</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co"># Compute p-value</span>
<span class="fu"><a href="../reference/getPValue.html">getPValue</a></span><span class="op">(</span>X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">X__</span><span class="op">)</span>, block_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X__</span><span class="op">)</span><span class="op">)</span>, largeP <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; [1] 9.398319e-61</span></code></pre></div>
<p>We obtain a <span class="math inline">\(p\)</span>-value that is essentially <span class="math inline">\(0\)</span>. In fact, we found that our test reports zero <span class="math inline">\(p\)</span>-values for each cluster. This is true even when we use TAD membership identified by <strong>TopDom</strong>, which we do not show here.</p>
</div>
<div id="investigating-homogeneity-of-subclusters" class="section level2">
<h2 class="hasAnchor">
<a href="#investigating-homogeneity-of-subclusters" class="anchor"></a>Investigating Homogeneity of Subclusters</h2>
<p>Perhaps the zero <span class="math inline">\(p\)</span>-values aren‚Äôt surprising, given that all clusters, except for Cluster 8, contain multiple cell types. We proceed to analyze subclusters consisting of unique cell types. We run our exchangeability test using the same grouping of peaks by TAD membership. Here, we find that all but three subclusters are non-exchangeable. The three exchangeable subclusters are printed below.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Show the three exchangeable subclusters</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com">data.table</a></span><span class="op">)</span>
<span class="va">domain_caller_results</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">rel_dir</span>, <span class="st">"/domainCaller_p_value_rm1s_df.txt"</span><span class="op">)</span><span class="op">)</span>
<span class="va">domain_caller_results</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/subset.data.table.html">subset</a></span><span class="op">(</span><span class="va">P_VALUE</span> <span class="op">&lt;</span> <span class="fl">1</span> <span class="op">&amp;</span> <span class="va">P_VALUE</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="co">#&gt;    CLUSTER  CELL_TYPE SAMPLE_SIZE   P_VALUE</span>
<span class="co">#&gt; 1:       0 CD4 Memory           6 0.2891715</span>
<span class="co">#&gt; 2:       0  CD4 Naive           5 0.1785740</span>
<span class="co">#&gt; 3:       6 CD4 Memory           7 0.4547475</span></code></pre></div>
<p>To summarize our results thus far, our test provides evidence of heterogeneity, not just in the clusters but also among subclusters consisting of a particular cell type. This is so even after accounting for correlations induced on peak read counts by the 3D structure (TADs) of the genome. Alternatively, if we assert that the clusters and subclusters are exchangeable, then grouping peaks by common TAD membership does not completely capture the complex dependencies between peaks.</p>
<p>We pursue these two directions of analyses below.</p>
</div>
<div id="what-drives-heterogeneity-in-clusters" class="section level2">
<h2 class="hasAnchor">
<a href="#what-drives-heterogeneity-in-clusters" class="anchor"></a>What Drives Heterogeneity in Clusters?</h2>
<p>Given that all <span class="math inline">\(15\)</span> clusters and most of their subclusters appear heterogeneous even after accounting for chromatin structure-induced dependencies, we shall attempt to explain what drives the heterogeneity. We consider a few <em>statistical</em> explanations. Each explanation can potentially be translated into some biological interpretation, and we defer the translation and interpretation to biologists who are more familiar with the subject-matter.</p>
<ul>
<li>
<em>Sparse peaks</em>. The normalized peaks, like the original counts matrix, are highly sparse. For example, when we compute the fraction of zero normalized reads per locus (column feature) for Cluster 11 cells, we find that about <span class="math inline">\(80\)</span>% of all loci (<span class="math inline">\(6769\)</span> out of <span class="math inline">\(8555\)</span>) have <span class="math inline">\(0.9\)</span> of their normalized reads equal to <span class="math inline">\(0\)</span>. Sparsity corresponds to the presence of non-zero values for a select few individuals, and contributes to the overall heterogeneity. This is analogous to the population genetic setting whereby rare variants are localized to a few individuals, typically related by recent ancestry. These individuals would thus form a subclade within the entire sample genealogy, breaking exchangeability.<br>
</li>
<li>
<em>Row sum variability</em>. There are two types of heterogeneities, one to do with differences in row sums of a sample and another to do with the patterning of zeros and non-zero values even if row sums are not too far apart. The first type of heterogeneity describes differences in the total value of non-zero peak read counts, across all measured peaks, between individual cells. The second type of heterogeneity describes differences in the configuration of non-zero peak read counts; for instance certain cells may have more non-zero peak read counts at a region while other samples may have more non-zero peak read counts at another region, despite the non-zero peak read counts across the genome being not too different.</li>
</ul>
<p>Here, we investigate the first point by dropping peaks that based on the fraction of non-zero normalized reads, and defer investigation of the second point to future work.</p>
<p>For example, we consider Cluster 12 (<span class="math inline">\(60\)</span> cells) and focus on the subcluster consisting of <span class="math inline">\(59\)</span> NK cells (one cell is a CD8 Naive). Note this subcluster is non-exchangeable when we run <strong>flinty</strong> at <span class="math inline">\(\alpha=0.05\)</span>. We find that a large proportion of peaks is sparse.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load and visualize peak statistics</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"cluster_12_pbmc_data.RData"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"090721_scATAC-seq_annotation_cluster.RData"</span><span class="op">)</span>
<span class="co"># Get peak locations</span>
<span class="va">peak_locs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">extracted_matrix</span><span class="op">)</span>
<span class="co"># Extract chromosome number</span>
<span class="va">chr_block_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">peak_locs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/substr.html">substring</a></span><span class="op">(</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/tstrsplit.html">strsplit</a></span><span class="op">(</span><span class="va">x</span>, split <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="fl">4</span><span class="op">)</span><span class="op">}</span><span class="op">)</span>
<span class="co"># Extract position of all chr1 chromosomes</span>
<span class="va">last_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="st">"2"</span>,<span class="va">chr_block_labels</span><span class="op">)</span>
<span class="va">last_id</span> <span class="op">&lt;-</span> <span class="va">last_id</span> <span class="op">-</span> <span class="fl">1</span>
<span class="va">X</span> <span class="op">&lt;-</span> <span class="va">extracted_matrix</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="va">last_id</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">each_peaks_tad_membership</span><span class="op">$</span><span class="va">LABEL</span>
<span class="co"># Delete the non-TAD classified features</span>
<span class="va">X_</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">each_peaks_tad_membership</span><span class="op">$</span><span class="va">LABEL</span> <span class="op">!=</span> <span class="op">-</span><span class="fl">99</span><span class="op">)</span><span class="op">]</span>
<span class="va">X__</span> <span class="op">&lt;-</span> <span class="va">X_</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X_</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span><span class="op">]</span>
<span class="co"># Filter out the single CD8 naive</span>
<span class="va">cluster_meta</span> <span class="op">&lt;-</span> <span class="va">combined_metadata</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">combined_metadata</span><span class="op">$</span><span class="va">umap_cluster_labels</span><span class="op">==</span><span class="fl">12</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">X_subclust</span> <span class="op">&lt;-</span> <span class="va">X__</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cluster_meta</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cluster_meta</span><span class="op">$</span><span class="va">pred_cell_type</span> <span class="op">==</span> <span class="st">"NK cell"</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span>,<span class="op">]</span>

<span class="co"># View histogram of feature sparsity (% zero normalized peak reads)</span>
<span class="va">col_sparse_ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/as.matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">X_subclust</span><span class="op">)</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">x</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">}</span><span class="op">)</span> 
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">col_sparse_ratio</span>, main <span class="op">=</span> <span class="st">"Histogram of Peak Sparsity"</span>, xlab <span class="op">=</span> <span class="st">"Peak Sparsity"</span><span class="op">)</span></code></pre></div>
<p><img src="ex_data/scATAC-seq/subcluster_1-1.png" style="width:80.0%"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">col_sparse_ratio</span> <span class="op">&lt;</span> <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.1429788</span>

<span class="co"># View plot of peak variance against peak sparsity</span>
<span class="va">col_variances</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">X_subclust</span>,<span class="fl">2</span>,<span class="va">var</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">col_sparse_ratio</span>, y <span class="op">=</span> <span class="va">col_variances</span>, 
     main <span class="op">=</span> <span class="st">"Peak Variance vs Peak Sparsity"</span>, 
     xlab <span class="op">=</span> <span class="st">"Peak Sparsity"</span>, 
     ylab <span class="op">=</span> <span class="st">"Peak Variance"</span><span class="op">)</span></code></pre></div>
<p><img src="ex_data/scATAC-seq/subcluster_1-2.png" style="width:80.0%"></p>
<p>To see if feature sparsity contributes to heterogeneity, we remove peaks from each group that are too sparse and try running our exchangeability test once more. After several iterations of filtering peaks, we find that additionally removing peaks that have sparsity <span class="math inline">\(\leqslant 0.7\)</span> leads to subcluster exchangeability.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Run test to explain contribution to heterogeneity</span>
<span class="co"># Remove sparse peaks</span>
<span class="va">X_non_sparse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/as.matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">X_subclust</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">col_sparse_ratio</span> <span class="op">&lt;</span> <span class="fl">0.8</span> <span class="op">&amp;</span> <span class="va">col_sparse_ratio</span> <span class="op">&gt;</span> <span class="fl">0.7</span><span class="op">)</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X_non_sparse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X_non_sparse</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Check number of independent blocks left after removal</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X_non_sparse</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> 
<span class="co">#&gt; [1] 148</span>
<span class="co"># large enough for asymptotic test to be used</span>

<span class="co"># Run flinty</span>
<span class="fu"><a href="../reference/getPValue.html">getPValue</a></span><span class="op">(</span><span class="va">X_non_sparse</span>, block_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X_non_sparse</span><span class="op">)</span><span class="op">)</span>, largeP <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.1579093</span></code></pre></div>
</div>
<div id="under-what-dependency-structure-of-peaks-are-clusters-exchangeable" class="section level2">
<h2 class="hasAnchor">
<a href="#under-what-dependency-structure-of-peaks-are-clusters-exchangeable" class="anchor"></a>Under what Dependency Structure of Peaks Are Clusters Exchangeable?</h2>
<p>As mentioned earlier, we may also assert that the clusters are exchangeable. If so, we must explain how the features are dependent for each cluster, such that under the dependency structure the cluster is exchangeable or homogeneous.</p>
<p>Let us consider Cluster 8, which is a community of <span class="math inline">\(244\)</span> pre-B cells. Like our analysis of Cluster 12 above, we start with visualizing some peak statistics, including the distributions of peak sparsity and peak variances.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load and visualize peak statistics</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"cluster_8_pbmc_data.RData"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"090721_scATAC-seq_annotation_cluster.RData"</span><span class="op">)</span>

<span class="co"># Get peak locations</span>
<span class="va">peak_locs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">extracted_matrix</span><span class="op">)</span>

<span class="co"># Extract chromosome number</span>
<span class="va">chr_block_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">peak_locs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/substr.html">substring</a></span><span class="op">(</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/tstrsplit.html">strsplit</a></span><span class="op">(</span><span class="va">x</span>, split <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="fl">4</span><span class="op">)</span><span class="op">}</span><span class="op">)</span>

<span class="co"># Extract position of all chr1 chromosomes</span>
<span class="va">last_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="st">"2"</span>,<span class="va">chr_block_labels</span><span class="op">)</span>
<span class="va">last_id</span> <span class="op">&lt;-</span> <span class="va">last_id</span> <span class="op">-</span> <span class="fl">1</span>
<span class="va">X</span> <span class="op">&lt;-</span> <span class="va">extracted_matrix</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="va">last_id</span><span class="op">]</span>

<span class="co"># View histogram of feature sparsity (% zero normalized peak reads)</span>
<span class="va">col_sparse_ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/as.matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">x</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">}</span><span class="op">)</span> 
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">col_sparse_ratio</span>, main <span class="op">=</span> <span class="st">"Histogram of Peak Sparsity"</span>, xlab <span class="op">=</span> <span class="st">"Peak Sparsity"</span><span class="op">)</span></code></pre></div>
<p><img src="ex_data/scATAC-seq/cluster_1-1.png" style="width:80.0%"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">col_sparse_ratio</span> <span class="op">&lt;</span> <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.1246055</span>

<span class="co"># View plot of peak variance against peak sparsity</span>
<span class="va">col_variances</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">X</span>,<span class="fl">2</span>,<span class="va">var</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">col_sparse_ratio</span>, y <span class="op">=</span> <span class="va">col_variances</span>, 
     main <span class="op">=</span> <span class="st">"Peak Variance vs Peak Sparsity"</span>, 
     xlab <span class="op">=</span> <span class="st">"Peak Sparsity"</span>, 
     ylab <span class="op">=</span> <span class="st">"Peak Variance"</span><span class="op">)</span></code></pre></div>
<p><img src="ex_data/scATAC-seq/cluster_1-2.png" style="width:80.0%"></p>
<p>We now try to group the peaks based on correlations. We perform hierarchical clustering with Ward linkage and distance matrix constructed from peak-peak correlations. (See p.¬†21 of <a href="https://www.jstatsoft.org/article/view/v061i06">Charrad et al., 2014</a> for an explanation of the Ward linkage.) To choose the optimal number of clusters, which correspond to groups of peaks, we use the Dindex metric. (See p.¬†14 of <a href="https://www.jstatsoft.org/article/view/v061i06">Charrad et al., 2014</a> for an explanation of how the Dindex works.) We find that <span class="math inline">\(K=46\)</span> clusters is optimal. Thus, we run our test again, with the peaks now grouped into <span class="math inline">\(46\)</span> blocks based on the clustering approach just described.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create distance matrix based on pairwise correlations</span>
<span class="va">peak_cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/as.matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span>
<span class="va">na_sums</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">peak_cor</span>,<span class="fl">2</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">}</span><span class="op">)</span>
<span class="va">na_peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">na_sums</span> <span class="op">==</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">peak_cor</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="va">peak_cor_na_rm</span> <span class="op">&lt;-</span> <span class="va">peak_cor</span><span class="op">[</span><span class="op">-</span><span class="va">na_peaks</span>,<span class="op">-</span><span class="va">na_peaks</span><span class="op">]</span>
<span class="va">peak_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">peak_cor_na_rm</span><span class="op">)</span> 

<span class="co"># Subset X</span>
<span class="va">X_</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/as.matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">[</span>,<span class="op">!</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">na_peaks</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>

<span class="co"># Remove peak correlation matrix</span>
<span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">peak_cor</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">peak_cor_na_rm</span><span class="op">)</span>

<span class="co"># Perform hierarchical clustering</span>
<span class="va">peak_tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="va">peak_dist</span>, method<span class="op">=</span><span class="st">"ward.D"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">peak_dist</span><span class="op">)</span>

<span class="co"># Clear garbage</span>
<span class="fu"><a href="https://rdrr.io/r/base/gc.html">gc</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt;             used   (Mb) gc trigger   (Mb) limit (Mb)  max used   (Mb)</span>
<span class="co">#&gt; Ncells  12172693  650.1   32126820 1715.8         NA  32126820 1715.8</span>
<span class="co">#&gt; Vcells 309536399 2361.6  693332412 5289.8      16384 693332222 5289.8</span>

<span class="co"># Create groups of features</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">dendextend</span><span class="op">)</span>
<span class="va">peak_dend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span><span class="op">(</span><span class="va">peak_tree</span><span class="op">)</span>

<span class="co">#library(NbClust) # for NbClust to work</span>
<span class="co"># This tuning step takes ~0.5h</span>
<span class="co">#res &lt;- NbClust(data.matrix(X_),</span>
<span class="co">#               distance = "euclidean", </span>
<span class="co">#               min.nc = 10, </span>
<span class="co">#               max.nc = 50, </span>
<span class="co">#               method = "ward.D2",</span>
<span class="co">#               index = "dindex")</span>
<span class="co">#-diff(res$All.index) #  minimized at q = 47, so N_opt = (q - 1) = 46</span>
<span class="co">#which(-diff(res$All.index) == min(-diff(res$All.index)))</span>

<span class="co"># View sizes of each group</span>
<span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html">cutree</a></span><span class="op">(</span><span class="va">peak_dend</span>, k <span class="op">=</span> <span class="fl">46</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span>
<span class="co">#&gt; clusters</span>
<span class="co">#&gt;    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16 </span>
<span class="co">#&gt;   63 5075   49   63   22   34   27   49   58   21   47   21   60   56   31   32 </span>
<span class="co">#&gt;   17   18   19   20   21   22   23   24   25   26   27   28   29   30   31   32 </span>
<span class="co">#&gt;   35  134   48   41   33   21   34   57   48   35   72   40   83   47   41   47 </span>
<span class="co">#&gt;   33   34   35   36   37   38   39   40   41   42   43   44   45   46 </span>
<span class="co">#&gt;   32   30   35   79   47   44   29   31   48   25   17   30   24   36</span>

<span class="co"># Group peaks into clusters</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X_</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">clusters</span>

<span class="co"># Run test of exchangeability</span>
<span class="fu"><a href="../reference/getPValue.html">getPValue</a></span><span class="op">(</span><span class="va">X_</span>, block_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">X_</span><span class="op">)</span><span class="op">)</span>, largeP <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.3087334</span></code></pre></div>
<p>We see that under this grouping of peaks and at level <span class="math inline">\(\alpha=0.05\)</span>, the sample is exchangeable.</p>
<p>To interpret this finding and the associated grouping, we inspect how peaks within the same group are distributed across Chr 1.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Inspect with-group peaks' physical location distribution</span>
<span class="va">phys_loc_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>CHR <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span>,
                          START <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/tstrsplit.html">strsplit</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"-"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">}</span><span class="op">)</span><span class="op">)</span>,
                          END <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/tstrsplit.html">strsplit</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"-"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">}</span><span class="op">)</span><span class="op">)</span>,
                          GROUP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Save files for chromoMap to read </span>
<span class="co">#write.table(phys_loc_df[,c(1,2,3)], "chr_file.txt", col.names = FALSE, row.names = FALSE)</span>
<span class="co">#write.table(phys_loc_df[,c(1,2,3)], "chr_file.txt", col.names = FALSE, row.names = FALSE)</span>
<span class="co">#phys_loc_df %&gt;% </span>
<span class="co">#  mutate(ANNO = paste0("Group", GROUP)) %&gt;%</span>
<span class="co">#  select(ANNO, CHR, START, END) %&gt;%</span>
<span class="co">#  write.table(file = "anno_file.txt", col.names = FALSE, row.names = FALSE)</span>

<span class="va">phys_loc_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"anno_file.txt"</span>, col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CLUSTER"</span>,<span class="st">"CHR"</span>,<span class="st">"START"</span>,<span class="st">"END"</span><span class="op">)</span><span class="op">)</span>
<span class="va">phys_loc_df</span> <span class="op">&lt;-</span> <span class="va">phys_loc_df</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>GROUP <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">CLUSTER</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/substr.html">substring</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">6</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="op">)</span><span class="op">}</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="va">CHR</span>,<span class="va">START</span>,<span class="va">END</span>,<span class="va">GROUP</span><span class="op">)</span>
<span class="co">#&gt; Error in (function (classes, fdef, mtable) : unable to find an inherited method for function 'select' for signature '"data.frame"'</span>

<span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">phys_loc_df</span>,
      <span class="fl">1</span>,
      <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html">data.table</a></span><span class="op">(</span>CHR <span class="op">=</span> <span class="va">x</span><span class="op">[</span><span class="st">"CHR"</span><span class="op">]</span>, BP <span class="op">=</span> <span class="va">x</span><span class="op">[</span><span class="st">"START"</span><span class="op">]</span><span class="op">:</span><span class="va">x</span><span class="op">[</span><span class="st">"END"</span><span class="op">]</span>, GROUP <span class="op">=</span> <span class="va">x</span><span class="op">[</span><span class="st">"GROUP"</span><span class="op">]</span><span class="op">)</span>
<span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/rbindlist.html">rbindlist</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#d7191c"</span>, <span class="st">"#fdae61"</span>, <span class="st">"#abd9e9"</span>, <span class="st">"#2c7bb6"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">46</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CHR</span>, y <span class="op">=</span> <span class="va">BP</span>, fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">GROUP</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">colors</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_x_discrete</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"Chromosome 1"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Physical Location (bp)"</span><span class="op">)</span>  <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">ylim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">565550</span>, <span class="fl">249219853</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"Group"</span><span class="op">)</span> </code></pre></div>
<p><img src="ex_data/scATAC-seq/cluster_3-1.png" style="width:80.0%"></p>
</div>
<div id="implications-for-statistical-modeling-of-scatac-seq-peaks" class="section level2">
<h2 class="hasAnchor">
<a href="#implications-for-statistical-modeling-of-scatac-seq-peaks" class="anchor"></a>Implications for Statistical Modeling of scATAC-seq Peaks</h2>
<p>Given the clusters found above do not group neatly by physical location, we interpret this as evidence that ‚Äúlong-range‚Äù peak interactions manifesting in correlations in the data, rather than grouping of peaks into contiguous blocks, must be accounted for in the modeling of scATAC-seq peaks, if the cluster is exchangeable or unstratified. There is some biological basis for this: there are almost certainly trans-acting factors that co-regulate multiple peaks across different TADs. Such co-regulation would likely induce dependencies between the peaks that are not captured by the TAD structure alone.</p>
<p>A broader implication of this analysis is that single cell ATAC peaks cannot simply be grouped into contiguous blocks (with contiguity based on physical location and TAD membership), as we saw earlier. Adequate <em>statistical modeling</em> of scATAC-seq data should go beyond simplistic blocking. This is unlike the case in statistical and population genetic models of evolution, where single nucleotide polymorphisms are grouped into blocks based on high within-block linkage disequilibrium (LD).</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Alan Aw.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
